// @livekit/components-react/dist/contexts-BIrOxv6y.mjs
import { setLogLevel as Wn, LogLevel as yt, setLogExtension as Bn, RoomEvent as g, ParticipantEvent as w, Room as wt, Track as D, TrackEvent as Vt, compareVersions as Vn } from "livekit-client";
import * as M from "react";
const tt = Math.min, le = Math.max, Me = Math.round, q = (e) => ({
  x: e,
  y: e
}), Hn = {
  left: "right",
  right: "left",
  bottom: "top",
  top: "bottom"
}, zn = {
  start: "end",
  end: "start"
};
function xt(e, t, n) {
  return le(e, tt(t, n));
}
function We(e, t) {
  return typeof e == "function" ? e(t) : e;
}
function se(e) {
  return e.split("-")[0];
}
function Be(e) {
  return e.split("-")[1];
}
function Ht(e) {
  return e === "x" ? "y" : "x";
}
function zt(e) {
  return e === "y" ? "height" : "width";
}
function pe(e) {
  return ["top", "bottom"].includes(se(e)) ? "y" : "x";
}
function Yt(e) {
  return Ht(pe(e));
}
function Yn(e, t, n) {
  n === void 0 && (n = !1);
  const r = Be(e), i = Yt(e), o = zt(i);
  let s = i === "x" ? r === (n ? "end" : "start") ? "right" : "left" : r === "start" ? "bottom" : "top";
  return t.reference[o] > t.floating[o] && (s = De(s)), [s, De(s)];
}
function Kn(e) {
  const t = De(e);
  return [nt(e), t, nt(t)];
}
function nt(e) {
  return e.replace(/start|end/g, (t) => zn[t]);
}
function Gn(e, t, n) {
  const r = ["left", "right"], i = ["right", "left"], o = ["top", "bottom"], s = ["bottom", "top"];
  switch (e) {
    case "top":
    case "bottom":
      return n ? t ? i : r : t ? r : i;
    case "left":
    case "right":
      return t ? o : s;
    default:
      return [];
  }
}
function qn(e, t, n, r) {
  const i = Be(e);
  let o = Gn(se(e), n === "start", r);
  return i && (o = o.map((s) => s + "-" + i), t && (o = o.concat(o.map(nt)))), o;
}
function De(e) {
  return e.replace(/left|right|bottom|top/g, (t) => Hn[t]);
}
function Qn(e) {
  return {
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
    ...e
  };
}
function Jn(e) {
  return typeof e != "number" ? Qn(e) : {
    top: e,
    right: e,
    bottom: e,
    left: e
  };
}
function Re(e) {
  const {
    x: t,
    y: n,
    width: r,
    height: i
  } = e;
  return {
    width: r,
    height: i,
    top: n,
    left: t,
    right: t + r,
    bottom: n + i,
    x: t,
    y: n
  };
}
function St(e, t, n) {
  let {
    reference: r,
    floating: i
  } = e;
  const o = pe(t), s = Yt(t), a = zt(s), c = se(t), u = o === "y", l = r.x + r.width / 2 - i.width / 2, f = r.y + r.height / 2 - i.height / 2, h = r[a] / 2 - i[a] / 2;
  let d;
  switch (c) {
    case "top":
      d = {
        x: l,
        y: r.y - i.height
      };
      break;
    case "bottom":
      d = {
        x: l,
        y: r.y + r.height
      };
      break;
    case "right":
      d = {
        x: r.x + r.width,
        y: f
      };
      break;
    case "left":
      d = {
        x: r.x - i.width,
        y: f
      };
      break;
    default:
      d = {
        x: r.x,
        y: r.y
      };
  }
  switch (Be(t)) {
    case "start":
      d[s] -= h * (n && u ? -1 : 1);
      break;
    case "end":
      d[s] += h * (n && u ? -1 : 1);
      break;
  }
  return d;
}
const Xn = async (e, t, n) => {
  const {
    placement: r = "bottom",
    strategy: i = "absolute",
    middleware: o = [],
    platform: s
  } = n, a = o.filter(Boolean), c = await (s.isRTL == null ? void 0 : s.isRTL(t));
  let u = await s.getElementRects({
    reference: e,
    floating: t,
    strategy: i
  }), {
    x: l,
    y: f
  } = St(u, r, c), h = r, d = {}, m = 0;
  for (let p = 0; p < a.length; p++) {
    const {
      name: b,
      fn: v
    } = a[p], {
      x: S,
      y: E,
      data: A,
      reset: y
    } = await v({
      x: l,
      y: f,
      initialPlacement: r,
      placement: h,
      strategy: i,
      middlewareData: d,
      rects: u,
      platform: s,
      elements: {
        reference: e,
        floating: t
      }
    });
    l = S ?? l, f = E ?? f, d = {
      ...d,
      [b]: {
        ...d[b],
        ...A
      }
    }, y && m <= 50 && (m++, typeof y == "object" && (y.placement && (h = y.placement), y.rects && (u = y.rects === !0 ? await s.getElementRects({
      reference: e,
      floating: t,
      strategy: i
    }) : y.rects), {
      x: l,
      y: f
    } = St(u, h, c)), p = -1);
  }
  return {
    x: l,
    y: f,
    placement: h,
    strategy: i,
    middlewareData: d
  };
};
async function Kt(e, t) {
  var n;
  t === void 0 && (t = {});
  const {
    x: r,
    y: i,
    platform: o,
    rects: s,
    elements: a,
    strategy: c
  } = e, {
    boundary: u = "clippingAncestors",
    rootBoundary: l = "viewport",
    elementContext: f = "floating",
    altBoundary: h = !1,
    padding: d = 0
  } = We(t, e), m = Jn(d), b = a[h ? f === "floating" ? "reference" : "floating" : f], v = Re(await o.getClippingRect({
    element: (n = await (o.isElement == null ? void 0 : o.isElement(b))) == null || n ? b : b.contextElement || await (o.getDocumentElement == null ? void 0 : o.getDocumentElement(a.floating)),
    boundary: u,
    rootBoundary: l,
    strategy: c
  })), S = f === "floating" ? {
    x: r,
    y: i,
    width: s.floating.width,
    height: s.floating.height
  } : s.reference, E = await (o.getOffsetParent == null ? void 0 : o.getOffsetParent(a.floating)), A = await (o.isElement == null ? void 0 : o.isElement(E)) ? await (o.getScale == null ? void 0 : o.getScale(E)) || {
    x: 1,
    y: 1
  } : {
    x: 1,
    y: 1
  }, y = Re(o.convertOffsetParentRelativeRectToViewportRelativeRect ? await o.convertOffsetParentRelativeRectToViewportRelativeRect({
    elements: a,
    rect: S,
    offsetParent: E,
    strategy: c
  }) : S);
  return {
    top: (v.top - y.top + m.top) / A.y,
    bottom: (y.bottom - v.bottom + m.bottom) / A.y,
    left: (v.left - y.left + m.left) / A.x,
    right: (y.right - v.right + m.right) / A.x
  };
}
const Zn = function(e) {
  return e === void 0 && (e = {}), {
    name: "flip",
    options: e,
    async fn(t) {
      var n, r;
      const {
        placement: i,
        middlewareData: o,
        rects: s,
        initialPlacement: a,
        platform: c,
        elements: u
      } = t, {
        mainAxis: l = !0,
        crossAxis: f = !0,
        fallbackPlacements: h,
        fallbackStrategy: d = "bestFit",
        fallbackAxisSideDirection: m = "none",
        flipAlignment: p = !0,
        ...b
      } = We(e, t);
      if ((n = o.arrow) != null && n.alignmentOffset)
        return {};
      const v = se(i), S = pe(a), E = se(a) === a, A = await (c.isRTL == null ? void 0 : c.isRTL(u.floating)), y = h || (E || !p ? [De(a)] : Kn(a)), x = m !== "none";
      !h && x && y.push(...qn(a, p, m, A));
      const O = [a, ...y], $ = await Kt(t, b), I = [];
      let z = ((r = o.flip) == null ? void 0 : r.overflows) || [];
      if (l && I.push($[v]), f) {
        const G = Yn(i, s, A);
        I.push($[G[0]], $[G[1]]);
      }
      if (z = [...z, {
        placement: i,
        overflows: I
      }], !I.every((G) => G <= 0)) {
        var T, L;
        const G = (((T = o.flip) == null ? void 0 : T.index) || 0) + 1, Oe = O[G];
        if (Oe)
          return {
            data: {
              index: G,
              overflows: z
            },
            reset: {
              placement: Oe
            }
          };
        let we = (L = z.filter((ue) => ue.overflows[0] <= 0).sort((ue, te) => ue.overflows[1] - te.overflows[1])[0]) == null ? void 0 : L.placement;
        if (!we)
          switch (d) {
            case "bestFit": {
              var oe;
              const ue = (oe = z.filter((te) => {
                if (x) {
                  const ne = pe(te.placement);
                  return ne === S || // Create a bias to the `y` side axis due to horizontal
                  // reading directions favoring greater width.
                  ne === "y";
                }
                return !0;
              }).map((te) => [te.placement, te.overflows.filter((ne) => ne > 0).reduce((ne, jn) => ne + jn, 0)]).sort((te, ne) => te[1] - ne[1])[0]) == null ? void 0 : oe[0];
              ue && (we = ue);
              break;
            }
            case "initialPlacement":
              we = a;
              break;
          }
        if (i !== we)
          return {
            reset: {
              placement: we
            }
          };
      }
      return {};
    }
  };
};
async function er(e, t) {
  const {
    placement: n,
    platform: r,
    elements: i
  } = e, o = await (r.isRTL == null ? void 0 : r.isRTL(i.floating)), s = se(n), a = Be(n), c = pe(n) === "y", u = ["left", "top"].includes(s) ? -1 : 1, l = o && c ? -1 : 1, f = We(t, e);
  let {
    mainAxis: h,
    crossAxis: d,
    alignmentAxis: m
  } = typeof f == "number" ? {
    mainAxis: f,
    crossAxis: 0,
    alignmentAxis: null
  } : {
    mainAxis: f.mainAxis || 0,
    crossAxis: f.crossAxis || 0,
    alignmentAxis: f.alignmentAxis
  };
  return a && typeof m == "number" && (d = a === "end" ? m * -1 : m), c ? {
    x: d * l,
    y: h * u
  } : {
    x: h * u,
    y: d * l
  };
}
const tr = function(e) {
  return e === void 0 && (e = 0), {
    name: "offset",
    options: e,
    async fn(t) {
      var n, r;
      const {
        x: i,
        y: o,
        placement: s,
        middlewareData: a
      } = t, c = await er(t, e);
      return s === ((n = a.offset) == null ? void 0 : n.placement) && (r = a.arrow) != null && r.alignmentOffset ? {} : {
        x: i + c.x,
        y: o + c.y,
        data: {
          ...c,
          placement: s
        }
      };
    }
  };
}, nr = function(e) {
  return e === void 0 && (e = {}), {
    name: "shift",
    options: e,
    async fn(t) {
      const {
        x: n,
        y: r,
        placement: i
      } = t, {
        mainAxis: o = !0,
        crossAxis: s = !1,
        limiter: a = {
          fn: (b) => {
            let {
              x: v,
              y: S
            } = b;
            return {
              x: v,
              y: S
            };
          }
        },
        ...c
      } = We(e, t), u = {
        x: n,
        y: r
      }, l = await Kt(t, c), f = pe(se(i)), h = Ht(f);
      let d = u[h], m = u[f];
      if (o) {
        const b = h === "y" ? "top" : "left", v = h === "y" ? "bottom" : "right", S = d + l[b], E = d - l[v];
        d = xt(S, d, E);
      }
      if (s) {
        const b = f === "y" ? "top" : "left", v = f === "y" ? "bottom" : "right", S = m + l[b], E = m - l[v];
        m = xt(S, m, E);
      }
      const p = a.fn({
        ...t,
        [h]: d,
        [f]: m
      });
      return {
        ...p,
        data: {
          x: p.x - n,
          y: p.y - r,
          enabled: {
            [h]: o,
            [f]: s
          }
        }
      };
    }
  };
};
function Ve() {
  return typeof window < "u";
}
function be(e) {
  return Gt(e) ? (e.nodeName || "").toLowerCase() : "#document";
}
function j(e) {
  var t;
  return (e == null || (t = e.ownerDocument) == null ? void 0 : t.defaultView) || window;
}
function Z(e) {
  var t;
  return (t = (Gt(e) ? e.ownerDocument : e.document) || window.document) == null ? void 0 : t.documentElement;
}
function Gt(e) {
  return Ve() ? e instanceof Node || e instanceof j(e).Node : !1;
}
function Y(e) {
  return Ve() ? e instanceof Element || e instanceof j(e).Element : !1;
}
function Q(e) {
  return Ve() ? e instanceof HTMLElement || e instanceof j(e).HTMLElement : !1;
}
function Tt(e) {
  return !Ve() || typeof ShadowRoot > "u" ? !1 : e instanceof ShadowRoot || e instanceof j(e).ShadowRoot;
}
function Pe(e) {
  const {
    overflow: t,
    overflowX: n,
    overflowY: r,
    display: i
  } = K(e);
  return /auto|scroll|overlay|hidden|clip/.test(t + r + n) && !["inline", "contents"].includes(i);
}
function rr(e) {
  return ["table", "td", "th"].includes(be(e));
}
function He(e) {
  return [":popover-open", ":modal"].some((t) => {
    try {
      return e.matches(t);
    } catch {
      return !1;
    }
  });
}
function ut(e) {
  const t = lt(), n = Y(e) ? K(e) : e;
  return ["transform", "translate", "scale", "rotate", "perspective"].some((r) => n[r] ? n[r] !== "none" : !1) || (n.containerType ? n.containerType !== "normal" : !1) || !t && (n.backdropFilter ? n.backdropFilter !== "none" : !1) || !t && (n.filter ? n.filter !== "none" : !1) || ["transform", "translate", "scale", "rotate", "perspective", "filter"].some((r) => (n.willChange || "").includes(r)) || ["paint", "layout", "strict", "content"].some((r) => (n.contain || "").includes(r));
}
function ir(e) {
  let t = ie(e);
  for (; Q(t) && !he(t); ) {
    if (ut(t))
      return t;
    if (He(t))
      return null;
    t = ie(t);
  }
  return null;
}
function lt() {
  return typeof CSS > "u" || !CSS.supports ? !1 : CSS.supports("-webkit-backdrop-filter", "none");
}
function he(e) {
  return ["html", "body", "#document"].includes(be(e));
}
function K(e) {
  return j(e).getComputedStyle(e);
}
function ze(e) {
  return Y(e) ? {
    scrollLeft: e.scrollLeft,
    scrollTop: e.scrollTop
  } : {
    scrollLeft: e.scrollX,
    scrollTop: e.scrollY
  };
}
function ie(e) {
  if (be(e) === "html")
    return e;
  const t = (
    // Step into the shadow DOM of the parent of a slotted node.
    e.assignedSlot || // DOM Element detected.
    e.parentNode || // ShadowRoot detected.
    Tt(e) && e.host || // Fallback.
    Z(e)
  );
  return Tt(t) ? t.host : t;
}
function qt(e) {
  const t = ie(e);
  return he(t) ? e.ownerDocument ? e.ownerDocument.body : e.body : Q(t) && Pe(t) ? t : qt(t);
}
function Qt(e, t, n) {
  var r;
  t === void 0 && (t = []);
  const i = qt(e), o = i === ((r = e.ownerDocument) == null ? void 0 : r.body), s = j(i);
  return o ? (rt(s), t.concat(s, s.visualViewport || [], Pe(i) ? i : [], [])) : t.concat(i, Qt(i, []));
}
function rt(e) {
  return e.parent && Object.getPrototypeOf(e.parent) ? e.frameElement : null;
}
function Jt(e) {
  const t = K(e);
  let n = parseFloat(t.width) || 0, r = parseFloat(t.height) || 0;
  const i = Q(e), o = i ? e.offsetWidth : n, s = i ? e.offsetHeight : r, a = Me(n) !== o || Me(r) !== s;
  return a && (n = o, r = s), {
    width: n,
    height: r,
    $: a
  };
}
function Xt(e) {
  return Y(e) ? e : e.contextElement;
}
function fe(e) {
  const t = Xt(e);
  if (!Q(t))
    return q(1);
  const n = t.getBoundingClientRect(), {
    width: r,
    height: i,
    $: o
  } = Jt(t);
  let s = (o ? Me(n.width) : n.width) / r, a = (o ? Me(n.height) : n.height) / i;
  return (!s || !Number.isFinite(s)) && (s = 1), (!a || !Number.isFinite(a)) && (a = 1), {
    x: s,
    y: a
  };
}
const or = /* @__PURE__ */ q(0);
function Zt(e) {
  const t = j(e);
  return !lt() || !t.visualViewport ? or : {
    x: t.visualViewport.offsetLeft,
    y: t.visualViewport.offsetTop
  };
}
function sr(e, t, n) {
  return t === void 0 && (t = !1), !n || t && n !== j(e) ? !1 : t;
}
function Te(e, t, n, r) {
  t === void 0 && (t = !1), n === void 0 && (n = !1);
  const i = e.getBoundingClientRect(), o = Xt(e);
  let s = q(1);
  t && (r ? Y(r) && (s = fe(r)) : s = fe(e));
  const a = sr(o, n, r) ? Zt(o) : q(0);
  let c = (i.left + a.x) / s.x, u = (i.top + a.y) / s.y, l = i.width / s.x, f = i.height / s.y;
  if (o) {
    const h = j(o), d = r && Y(r) ? j(r) : r;
    let m = h, p = rt(m);
    for (; p && r && d !== m; ) {
      const b = fe(p), v = p.getBoundingClientRect(), S = K(p), E = v.left + (p.clientLeft + parseFloat(S.paddingLeft)) * b.x, A = v.top + (p.clientTop + parseFloat(S.paddingTop)) * b.y;
      c *= b.x, u *= b.y, l *= b.x, f *= b.y, c += E, u += A, m = j(p), p = rt(m);
    }
  }
  return Re({
    width: l,
    height: f,
    x: c,
    y: u
  });
}
function ft(e, t) {
  const n = ze(e).scrollLeft;
  return t ? t.left + n : Te(Z(e)).left + n;
}
function en(e, t, n) {
  n === void 0 && (n = !1);
  const r = e.getBoundingClientRect(), i = r.left + t.scrollLeft - (n ? 0 : (
    // RTL <body> scrollbar.
    ft(e, r)
  )), o = r.top + t.scrollTop;
  return {
    x: i,
    y: o
  };
}
function ar(e) {
  let {
    elements: t,
    rect: n,
    offsetParent: r,
    strategy: i
  } = e;
  const o = i === "fixed", s = Z(r), a = t ? He(t.floating) : !1;
  if (r === s || a && o)
    return n;
  let c = {
    scrollLeft: 0,
    scrollTop: 0
  }, u = q(1);
  const l = q(0), f = Q(r);
  if ((f || !f && !o) && ((be(r) !== "body" || Pe(s)) && (c = ze(r)), Q(r))) {
    const d = Te(r);
    u = fe(r), l.x = d.x + r.clientLeft, l.y = d.y + r.clientTop;
  }
  const h = s && !f && !o ? en(s, c, !0) : q(0);
  return {
    width: n.width * u.x,
    height: n.height * u.y,
    x: n.x * u.x - c.scrollLeft * u.x + l.x + h.x,
    y: n.y * u.y - c.scrollTop * u.y + l.y + h.y
  };
}
function cr(e) {
  return Array.from(e.getClientRects());
}
function ur(e) {
  const t = Z(e), n = ze(e), r = e.ownerDocument.body, i = le(t.scrollWidth, t.clientWidth, r.scrollWidth, r.clientWidth), o = le(t.scrollHeight, t.clientHeight, r.scrollHeight, r.clientHeight);
  let s = -n.scrollLeft + ft(e);
  const a = -n.scrollTop;
  return K(r).direction === "rtl" && (s += le(t.clientWidth, r.clientWidth) - i), {
    width: i,
    height: o,
    x: s,
    y: a
  };
}
function lr(e, t) {
  const n = j(e), r = Z(e), i = n.visualViewport;
  let o = r.clientWidth, s = r.clientHeight, a = 0, c = 0;
  if (i) {
    o = i.width, s = i.height;
    const u = lt();
    (!u || u && t === "fixed") && (a = i.offsetLeft, c = i.offsetTop);
  }
  return {
    width: o,
    height: s,
    x: a,
    y: c
  };
}
function fr(e, t) {
  const n = Te(e, !0, t === "fixed"), r = n.top + e.clientTop, i = n.left + e.clientLeft, o = Q(e) ? fe(e) : q(1), s = e.clientWidth * o.x, a = e.clientHeight * o.y, c = i * o.x, u = r * o.y;
  return {
    width: s,
    height: a,
    x: c,
    y: u
  };
}
function Et(e, t, n) {
  let r;
  if (t === "viewport")
    r = lr(e, n);
  else if (t === "document")
    r = ur(Z(e));
  else if (Y(t))
    r = fr(t, n);
  else {
    const i = Zt(e);
    r = {
      x: t.x - i.x,
      y: t.y - i.y,
      width: t.width,
      height: t.height
    };
  }
  return Re(r);
}
function tn(e, t) {
  const n = ie(e);
  return n === t || !Y(n) || he(n) ? !1 : K(n).position === "fixed" || tn(n, t);
}
function dr(e, t) {
  const n = t.get(e);
  if (n)
    return n;
  let r = Qt(e, []).filter((a) => Y(a) && be(a) !== "body"), i = null;
  const o = K(e).position === "fixed";
  let s = o ? ie(e) : e;
  for (; Y(s) && !he(s); ) {
    const a = K(s), c = ut(s);
    !c && a.position === "fixed" && (i = null), (o ? !c && !i : !c && a.position === "static" && !!i && ["absolute", "fixed"].includes(i.position) || Pe(s) && !c && tn(e, s)) ? r = r.filter((l) => l !== s) : i = a, s = ie(s);
  }
  return t.set(e, r), r;
}
function pr(e) {
  let {
    element: t,
    boundary: n,
    rootBoundary: r,
    strategy: i
  } = e;
  const s = [...n === "clippingAncestors" ? He(t) ? [] : dr(t, this._c) : [].concat(n), r], a = s[0], c = s.reduce((u, l) => {
    const f = Et(t, l, i);
    return u.top = le(f.top, u.top), u.right = tt(f.right, u.right), u.bottom = tt(f.bottom, u.bottom), u.left = le(f.left, u.left), u;
  }, Et(t, a, i));
  return {
    width: c.right - c.left,
    height: c.bottom - c.top,
    x: c.left,
    y: c.top
  };
}
function hr(e) {
  const {
    width: t,
    height: n
  } = Jt(e);
  return {
    width: t,
    height: n
  };
}
function vr(e, t, n) {
  const r = Q(t), i = Z(t), o = n === "fixed", s = Te(e, !0, o, t);
  let a = {
    scrollLeft: 0,
    scrollTop: 0
  };
  const c = q(0);
  if (r || !r && !o)
    if ((be(t) !== "body" || Pe(i)) && (a = ze(t)), r) {
      const h = Te(t, !0, o, t);
      c.x = h.x + t.clientLeft, c.y = h.y + t.clientTop;
    } else i && (c.x = ft(i));
  const u = i && !r && !o ? en(i, a) : q(0), l = s.left + a.scrollLeft - c.x - u.x, f = s.top + a.scrollTop - c.y - u.y;
  return {
    x: l,
    y: f,
    width: s.width,
    height: s.height
  };
}
function qe(e) {
  return K(e).position === "static";
}
function Ct(e, t) {
  if (!Q(e) || K(e).position === "fixed")
    return null;
  if (t)
    return t(e);
  let n = e.offsetParent;
  return Z(e) === n && (n = n.ownerDocument.body), n;
}
function nn(e, t) {
  const n = j(e);
  if (He(e))
    return n;
  if (!Q(e)) {
    let i = ie(e);
    for (; i && !he(i); ) {
      if (Y(i) && !qe(i))
        return i;
      i = ie(i);
    }
    return n;
  }
  let r = Ct(e, t);
  for (; r && rr(r) && qe(r); )
    r = Ct(r, t);
  return r && he(r) && qe(r) && !ut(r) ? n : r || ir(e) || n;
}
const mr = async function(e) {
  const t = this.getOffsetParent || nn, n = this.getDimensions, r = await n(e.floating);
  return {
    reference: vr(e.reference, await t(e.floating), e.strategy),
    floating: {
      x: 0,
      y: 0,
      width: r.width,
      height: r.height
    }
  };
};
function br(e) {
  return K(e).direction === "rtl";
}
const gr = {
  convertOffsetParentRelativeRectToViewportRelativeRect: ar,
  getDocumentElement: Z,
  getClippingRect: pr,
  getOffsetParent: nn,
  getElementRects: mr,
  getClientRects: cr,
  getDimensions: hr,
  getScale: fe,
  isElement: Y,
  isRTL: br
}, yr = tr, wr = nr, xr = Zn, Sr = (e, t, n) => {
  const r = /* @__PURE__ */ new Map(), i = {
    platform: gr,
    ...n
  }, o = {
    ...i.platform,
    _c: r
  };
  return Xn(e, t, {
    ...i,
    platform: o
  });
};
var _o = typeof globalThis < "u" ? globalThis : typeof window < "u" ? window : typeof global < "u" ? global : typeof self < "u" ? self : {};
function Tr(e) {
  return e && e.__esModule && Object.prototype.hasOwnProperty.call(e, "default") ? e.default : e;
}
var _e = { exports: {} }, Er = _e.exports, Pt;
function Cr() {
  return Pt || (Pt = 1, function(e) {
    (function(t, n) {
      e.exports ? e.exports = n() : t.log = n();
    })(Er, function() {
      var t = function() {
      }, n = "undefined", r = typeof window !== n && typeof window.navigator !== n && /Trident\/|MSIE /.test(window.navigator.userAgent), i = [
        "trace",
        "debug",
        "info",
        "warn",
        "error"
      ], o = {}, s = null;
      function a(p, b) {
        var v = p[b];
        if (typeof v.bind == "function")
          return v.bind(p);
        try {
          return Function.prototype.bind.call(v, p);
        } catch {
          return function() {
            return Function.prototype.apply.apply(v, [p, arguments]);
          };
        }
      }
      function c() {
        console.log && (console.log.apply ? console.log.apply(console, arguments) : Function.prototype.apply.apply(console.log, [console, arguments])), console.trace && console.trace();
      }
      function u(p) {
        return p === "debug" && (p = "log"), typeof console === n ? !1 : p === "trace" && r ? c : console[p] !== void 0 ? a(console, p) : console.log !== void 0 ? a(console, "log") : t;
      }
      function l() {
        for (var p = this.getLevel(), b = 0; b < i.length; b++) {
          var v = i[b];
          this[v] = b < p ? t : this.methodFactory(v, p, this.name);
        }
        if (this.log = this.debug, typeof console === n && p < this.levels.SILENT)
          return "No console available for logging";
      }
      function f(p) {
        return function() {
          typeof console !== n && (l.call(this), this[p].apply(this, arguments));
        };
      }
      function h(p, b, v) {
        return u(p) || f.apply(this, arguments);
      }
      function d(p, b) {
        var v = this, S, E, A, y = "loglevel";
        typeof p == "string" ? y += ":" + p : typeof p == "symbol" && (y = void 0);
        function x(T) {
          var L = (i[T] || "silent").toUpperCase();
          if (!(typeof window === n || !y)) {
            try {
              window.localStorage[y] = L;
              return;
            } catch {
            }
            try {
              window.document.cookie = encodeURIComponent(y) + "=" + L + ";";
            } catch {
            }
          }
        }
        function O() {
          var T;
          if (!(typeof window === n || !y)) {
            try {
              T = window.localStorage[y];
            } catch {
            }
            if (typeof T === n)
              try {
                var L = window.document.cookie, oe = encodeURIComponent(y), G = L.indexOf(oe + "=");
                G !== -1 && (T = /^([^;]+)/.exec(
                  L.slice(G + oe.length + 1)
                )[1]);
              } catch {
              }
            return v.levels[T] === void 0 && (T = void 0), T;
          }
        }
        function $() {
          if (!(typeof window === n || !y)) {
            try {
              window.localStorage.removeItem(y);
            } catch {
            }
            try {
              window.document.cookie = encodeURIComponent(y) + "=; expires=Thu, 01 Jan 1970 00:00:00 UTC";
            } catch {
            }
          }
        }
        function I(T) {
          var L = T;
          if (typeof L == "string" && v.levels[L.toUpperCase()] !== void 0 && (L = v.levels[L.toUpperCase()]), typeof L == "number" && L >= 0 && L <= v.levels.SILENT)
            return L;
          throw new TypeError("log.setLevel() called with invalid level: " + T);
        }
        v.name = p, v.levels = {
          TRACE: 0,
          DEBUG: 1,
          INFO: 2,
          WARN: 3,
          ERROR: 4,
          SILENT: 5
        }, v.methodFactory = b || h, v.getLevel = function() {
          return A ?? E ?? S;
        }, v.setLevel = function(T, L) {
          return A = I(T), L !== !1 && x(A), l.call(v);
        }, v.setDefaultLevel = function(T) {
          E = I(T), O() || v.setLevel(T, !1);
        }, v.resetLevel = function() {
          A = null, $(), l.call(v);
        }, v.enableAll = function(T) {
          v.setLevel(v.levels.TRACE, T);
        }, v.disableAll = function(T) {
          v.setLevel(v.levels.SILENT, T);
        }, v.rebuild = function() {
          if (s !== v && (S = I(s.getLevel())), l.call(v), s === v)
            for (var T in o)
              o[T].rebuild();
        }, S = I(
          s ? s.getLevel() : "WARN"
        );
        var z = O();
        z != null && (A = I(z)), l.call(v);
      }
      s = new d(), s.getLogger = function(b) {
        if (typeof b != "symbol" && typeof b != "string" || b === "")
          throw new TypeError("You must supply a name when creating a logger.");
        var v = o[b];
        return v || (v = o[b] = new d(
          b,
          s.methodFactory
        )), v;
      };
      var m = typeof window !== n ? window.log : void 0;
      return s.noConflict = function() {
        return typeof window !== n && window.log === s && (window.log = m), s;
      }, s.getLoggers = function() {
        return o;
      }, s.default = s, s;
    });
  }(_e)), _e.exports;
}
var Pr = Cr();
const Ar = /* @__PURE__ */ Tr(Pr);
var it = function(e, t) {
  return it = Object.setPrototypeOf || { __proto__: [] } instanceof Array && function(n, r) {
    n.__proto__ = r;
  } || function(n, r) {
    for (var i in r) Object.prototype.hasOwnProperty.call(r, i) && (n[i] = r[i]);
  }, it(e, t);
};
function ee(e, t) {
  if (typeof t != "function" && t !== null)
    throw new TypeError("Class extends value " + String(t) + " is not a constructor or null");
  it(e, t);
  function n() {
    this.constructor = e;
  }
  e.prototype = t === null ? Object.create(t) : (n.prototype = t.prototype, new n());
}
function Or(e, t, n, r) {
  function i(o) {
    return o instanceof n ? o : new n(function(s) {
      s(o);
    });
  }
  return new (n || (n = Promise))(function(o, s) {
    function a(l) {
      try {
        u(r.next(l));
      } catch (f) {
        s(f);
      }
    }
    function c(l) {
      try {
        u(r.throw(l));
      } catch (f) {
        s(f);
      }
    }
    function u(l) {
      l.done ? o(l.value) : i(l.value).then(a, c);
    }
    u((r = r.apply(e, t || [])).next());
  });
}
function rn(e, t) {
  var n = { label: 0, sent: function() {
    if (o[0] & 1) throw o[1];
    return o[1];
  }, trys: [], ops: [] }, r, i, o, s = Object.create((typeof Iterator == "function" ? Iterator : Object).prototype);
  return s.next = a(0), s.throw = a(1), s.return = a(2), typeof Symbol == "function" && (s[Symbol.iterator] = function() {
    return this;
  }), s;
  function a(u) {
    return function(l) {
      return c([u, l]);
    };
  }
  function c(u) {
    if (r) throw new TypeError("Generator is already executing.");
    for (; s && (s = 0, u[0] && (n = 0)), n; ) try {
      if (r = 1, i && (o = u[0] & 2 ? i.return : u[0] ? i.throw || ((o = i.return) && o.call(i), 0) : i.next) && !(o = o.call(i, u[1])).done) return o;
      switch (i = 0, o && (u = [u[0] & 2, o.value]), u[0]) {
        case 0:
        case 1:
          o = u;
          break;
        case 4:
          return n.label++, { value: u[1], done: !1 };
        case 5:
          n.label++, i = u[1], u = [0];
          continue;
        case 7:
          u = n.ops.pop(), n.trys.pop();
          continue;
        default:
          if (o = n.trys, !(o = o.length > 0 && o[o.length - 1]) && (u[0] === 6 || u[0] === 2)) {
            n = 0;
            continue;
          }
          if (u[0] === 3 && (!o || u[1] > o[0] && u[1] < o[3])) {
            n.label = u[1];
            break;
          }
          if (u[0] === 6 && n.label < o[1]) {
            n.label = o[1], o = u;
            break;
          }
          if (o && n.label < o[2]) {
            n.label = o[2], n.ops.push(u);
            break;
          }
          o[2] && n.ops.pop(), n.trys.pop();
          continue;
      }
      u = t.call(e, n);
    } catch (l) {
      u = [6, l], i = 0;
    } finally {
      r = o = 0;
    }
    if (u[0] & 5) throw u[1];
    return { value: u[0] ? u[1] : void 0, done: !0 };
  }
}
function ve(e) {
  var t = typeof Symbol == "function" && Symbol.iterator, n = t && e[t], r = 0;
  if (n) return n.call(e);
  if (e && typeof e.length == "number") return {
    next: function() {
      return e && r >= e.length && (e = void 0), { value: e && e[r++], done: !e };
    }
  };
  throw new TypeError(t ? "Object is not iterable." : "Symbol.iterator is not defined.");
}
function ae(e, t) {
  var n = typeof Symbol == "function" && e[Symbol.iterator];
  if (!n) return e;
  var r = n.call(e), i, o = [], s;
  try {
    for (; (t === void 0 || t-- > 0) && !(i = r.next()).done; ) o.push(i.value);
  } catch (a) {
    s = { error: a };
  } finally {
    try {
      i && !i.done && (n = r.return) && n.call(r);
    } finally {
      if (s) throw s.error;
    }
  }
  return o;
}
function me(e, t, n) {
  if (n || arguments.length === 2) for (var r = 0, i = t.length, o; r < i; r++)
    (o || !(r in t)) && (o || (o = Array.prototype.slice.call(t, 0, r)), o[r] = t[r]);
  return e.concat(o || Array.prototype.slice.call(t));
}
function de(e) {
  return this instanceof de ? (this.v = e, this) : new de(e);
}
function kr(e, t, n) {
  if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
  var r = n.apply(e, t || []), i, o = [];
  return i = Object.create((typeof AsyncIterator == "function" ? AsyncIterator : Object).prototype), a("next"), a("throw"), a("return", s), i[Symbol.asyncIterator] = function() {
    return this;
  }, i;
  function s(d) {
    return function(m) {
      return Promise.resolve(m).then(d, f);
    };
  }
  function a(d, m) {
    r[d] && (i[d] = function(p) {
      return new Promise(function(b, v) {
        o.push([d, p, b, v]) > 1 || c(d, p);
      });
    }, m && (i[d] = m(i[d])));
  }
  function c(d, m) {
    try {
      u(r[d](m));
    } catch (p) {
      h(o[0][3], p);
    }
  }
  function u(d) {
    d.value instanceof de ? Promise.resolve(d.value.v).then(l, f) : h(o[0][2], d);
  }
  function l(d) {
    c("next", d);
  }
  function f(d) {
    c("throw", d);
  }
  function h(d, m) {
    d(m), o.shift(), o.length && c(o[0][0], o[0][1]);
  }
}
function Lr(e) {
  if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
  var t = e[Symbol.asyncIterator], n;
  return t ? t.call(e) : (e = typeof ve == "function" ? ve(e) : e[Symbol.iterator](), n = {}, r("next"), r("throw"), r("return"), n[Symbol.asyncIterator] = function() {
    return this;
  }, n);
  function r(o) {
    n[o] = e[o] && function(s) {
      return new Promise(function(a, c) {
        s = e[o](s), i(a, c, s.done, s.value);
      });
    };
  }
  function i(o, s, a, c) {
    Promise.resolve(c).then(function(u) {
      o({ value: u, done: a });
    }, s);
  }
}
function C(e) {
  return typeof e == "function";
}
function dt(e) {
  var t = function(r) {
    Error.call(r), r.stack = new Error().stack;
  }, n = e(t);
  return n.prototype = Object.create(Error.prototype), n.prototype.constructor = n, n;
}
var Qe = dt(function(e) {
  return function(n) {
    e(this), this.message = n ? n.length + ` errors occurred during unsubscription:
` + n.map(function(r, i) {
      return i + 1 + ") " + r.toString();
    }).join(`
  `) : "", this.name = "UnsubscriptionError", this.errors = n;
  };
});
function $e(e, t) {
  if (e) {
    var n = e.indexOf(t);
    0 <= n && e.splice(n, 1);
  }
}
var Ae = function() {
  function e(t) {
    this.initialTeardown = t, this.closed = !1, this._parentage = null, this._finalizers = null;
  }
  return e.prototype.unsubscribe = function() {
    var t, n, r, i, o;
    if (!this.closed) {
      this.closed = !0;
      var s = this._parentage;
      if (s)
        if (this._parentage = null, Array.isArray(s))
          try {
            for (var a = ve(s), c = a.next(); !c.done; c = a.next()) {
              var u = c.value;
              u.remove(this);
            }
          } catch (p) {
            t = { error: p };
          } finally {
            try {
              c && !c.done && (n = a.return) && n.call(a);
            } finally {
              if (t) throw t.error;
            }
          }
        else
          s.remove(this);
      var l = this.initialTeardown;
      if (C(l))
        try {
          l();
        } catch (p) {
          o = p instanceof Qe ? p.errors : [p];
        }
      var f = this._finalizers;
      if (f) {
        this._finalizers = null;
        try {
          for (var h = ve(f), d = h.next(); !d.done; d = h.next()) {
            var m = d.value;
            try {
              At(m);
            } catch (p) {
              o = o ?? [], p instanceof Qe ? o = me(me([], ae(o)), ae(p.errors)) : o.push(p);
            }
          }
        } catch (p) {
          r = { error: p };
        } finally {
          try {
            d && !d.done && (i = h.return) && i.call(h);
          } finally {
            if (r) throw r.error;
          }
        }
      }
      if (o)
        throw new Qe(o);
    }
  }, e.prototype.add = function(t) {
    var n;
    if (t && t !== this)
      if (this.closed)
        At(t);
      else {
        if (t instanceof e) {
          if (t.closed || t._hasParent(this))
            return;
          t._addParent(this);
        }
        (this._finalizers = (n = this._finalizers) !== null && n !== void 0 ? n : []).push(t);
      }
  }, e.prototype._hasParent = function(t) {
    var n = this._parentage;
    return n === t || Array.isArray(n) && n.includes(t);
  }, e.prototype._addParent = function(t) {
    var n = this._parentage;
    this._parentage = Array.isArray(n) ? (n.push(t), n) : n ? [n, t] : t;
  }, e.prototype._removeParent = function(t) {
    var n = this._parentage;
    n === t ? this._parentage = null : Array.isArray(n) && $e(n, t);
  }, e.prototype.remove = function(t) {
    var n = this._finalizers;
    n && $e(n, t), t instanceof e && t._removeParent(this);
  }, e.EMPTY = function() {
    var t = new e();
    return t.closed = !0, t;
  }(), e;
}(), on = Ae.EMPTY;
function sn(e) {
  return e instanceof Ae || e && "closed" in e && C(e.remove) && C(e.add) && C(e.unsubscribe);
}
function At(e) {
  C(e) ? e() : e.unsubscribe();
}
var _r = {
  Promise: void 0
}, Ir = {
  setTimeout: function(e, t) {
    for (var n = [], r = 2; r < arguments.length; r++)
      n[r - 2] = arguments[r];
    return setTimeout.apply(void 0, me([e, t], ae(n)));
  },
  clearTimeout: function(e) {
    return clearTimeout(e);
  },
  delegate: void 0
};
function an(e) {
  Ir.setTimeout(function() {
    throw e;
  });
}
function Ne() {
}
function Ie(e) {
  e();
}
var pt = function(e) {
  ee(t, e);
  function t(n) {
    var r = e.call(this) || this;
    return r.isStopped = !1, n ? (r.destination = n, sn(n) && n.add(r)) : r.destination = Rr, r;
  }
  return t.create = function(n, r, i) {
    return new Ee(n, r, i);
  }, t.prototype.next = function(n) {
    this.isStopped || this._next(n);
  }, t.prototype.error = function(n) {
    this.isStopped || (this.isStopped = !0, this._error(n));
  }, t.prototype.complete = function() {
    this.isStopped || (this.isStopped = !0, this._complete());
  }, t.prototype.unsubscribe = function() {
    this.closed || (this.isStopped = !0, e.prototype.unsubscribe.call(this), this.destination = null);
  }, t.prototype._next = function(n) {
    this.destination.next(n);
  }, t.prototype._error = function(n) {
    try {
      this.destination.error(n);
    } finally {
      this.unsubscribe();
    }
  }, t.prototype._complete = function() {
    try {
      this.destination.complete();
    } finally {
      this.unsubscribe();
    }
  }, t;
}(Ae), Mr = function() {
  function e(t) {
    this.partialObserver = t;
  }
  return e.prototype.next = function(t) {
    var n = this.partialObserver;
    if (n.next)
      try {
        n.next(t);
      } catch (r) {
        ke(r);
      }
  }, e.prototype.error = function(t) {
    var n = this.partialObserver;
    if (n.error)
      try {
        n.error(t);
      } catch (r) {
        ke(r);
      }
    else
      ke(t);
  }, e.prototype.complete = function() {
    var t = this.partialObserver;
    if (t.complete)
      try {
        t.complete();
      } catch (n) {
        ke(n);
      }
  }, e;
}(), Ee = function(e) {
  ee(t, e);
  function t(n, r, i) {
    var o = e.call(this) || this, s;
    return C(n) || !n ? s = {
      next: n ?? void 0,
      error: r ?? void 0,
      complete: i ?? void 0
    } : s = n, o.destination = new Mr(s), o;
  }
  return t;
}(pt);
function ke(e) {
  an(e);
}
function Dr(e) {
  throw e;
}
var Rr = {
  closed: !0,
  next: Ne,
  error: Dr,
  complete: Ne
}, ht = function() {
  return typeof Symbol == "function" && Symbol.observable || "@@observable";
}();
function vt(e) {
  return e;
}
function $r(e) {
  return e.length === 0 ? vt : e.length === 1 ? e[0] : function(n) {
    return e.reduce(function(r, i) {
      return i(r);
    }, n);
  };
}
var k = function() {
  function e(t) {
    t && (this._subscribe = t);
  }
  return e.prototype.lift = function(t) {
    var n = new e();
    return n.source = this, n.operator = t, n;
  }, e.prototype.subscribe = function(t, n, r) {
    var i = this, o = Ur(t) ? t : new Ee(t, n, r);
    return Ie(function() {
      var s = i, a = s.operator, c = s.source;
      o.add(a ? a.call(o, c) : c ? i._subscribe(o) : i._trySubscribe(o));
    }), o;
  }, e.prototype._trySubscribe = function(t) {
    try {
      return this._subscribe(t);
    } catch (n) {
      t.error(n);
    }
  }, e.prototype.forEach = function(t, n) {
    var r = this;
    return n = Ot(n), new n(function(i, o) {
      var s = new Ee({
        next: function(a) {
          try {
            t(a);
          } catch (c) {
            o(c), s.unsubscribe();
          }
        },
        error: o,
        complete: i
      });
      r.subscribe(s);
    });
  }, e.prototype._subscribe = function(t) {
    var n;
    return (n = this.source) === null || n === void 0 ? void 0 : n.subscribe(t);
  }, e.prototype[ht] = function() {
    return this;
  }, e.prototype.pipe = function() {
    for (var t = [], n = 0; n < arguments.length; n++)
      t[n] = arguments[n];
    return $r(t)(this);
  }, e.prototype.toPromise = function(t) {
    var n = this;
    return t = Ot(t), new t(function(r, i) {
      var o;
      n.subscribe(function(s) {
        return o = s;
      }, function(s) {
        return i(s);
      }, function() {
        return r(o);
      });
    });
  }, e.create = function(t) {
    return new e(t);
  }, e;
}();
function Ot(e) {
  var t;
  return (t = e ?? _r.Promise) !== null && t !== void 0 ? t : Promise;
}
function Nr(e) {
  return e && C(e.next) && C(e.error) && C(e.complete);
}
function Ur(e) {
  return e && e instanceof pt || Nr(e) && sn(e);
}
function Fr(e) {
  return C(e == null ? void 0 : e.lift);
}
function F(e) {
  return function(t) {
    if (Fr(t))
      return t.lift(function(n) {
        try {
          return e(n, this);
        } catch (r) {
          this.error(r);
        }
      });
    throw new TypeError("Unable to lift unknown Observable type");
  };
}
function B(e, t, n, r, i) {
  return new jr(e, t, n, r, i);
}
var jr = function(e) {
  ee(t, e);
  function t(n, r, i, o, s, a) {
    var c = e.call(this, n) || this;
    return c.onFinalize = s, c.shouldUnsubscribe = a, c._next = r ? function(u) {
      try {
        r(u);
      } catch (l) {
        n.error(l);
      }
    } : e.prototype._next, c._error = o ? function(u) {
      try {
        o(u);
      } catch (l) {
        n.error(l);
      } finally {
        this.unsubscribe();
      }
    } : e.prototype._error, c._complete = i ? function() {
      try {
        i();
      } catch (u) {
        n.error(u);
      } finally {
        this.unsubscribe();
      }
    } : e.prototype._complete, c;
  }
  return t.prototype.unsubscribe = function() {
    var n;
    if (!this.shouldUnsubscribe || this.shouldUnsubscribe()) {
      var r = this.closed;
      e.prototype.unsubscribe.call(this), !r && ((n = this.onFinalize) === null || n === void 0 || n.call(this));
    }
  }, t;
}(pt), Wr = dt(function(e) {
  return function() {
    e(this), this.name = "ObjectUnsubscribedError", this.message = "object unsubscribed";
  };
}), J = function(e) {
  ee(t, e);
  function t() {
    var n = e.call(this) || this;
    return n.closed = !1, n.currentObservers = null, n.observers = [], n.isStopped = !1, n.hasError = !1, n.thrownError = null, n;
  }
  return t.prototype.lift = function(n) {
    var r = new kt(this, this);
    return r.operator = n, r;
  }, t.prototype._throwIfClosed = function() {
    if (this.closed)
      throw new Wr();
  }, t.prototype.next = function(n) {
    var r = this;
    Ie(function() {
      var i, o;
      if (r._throwIfClosed(), !r.isStopped) {
        r.currentObservers || (r.currentObservers = Array.from(r.observers));
        try {
          for (var s = ve(r.currentObservers), a = s.next(); !a.done; a = s.next()) {
            var c = a.value;
            c.next(n);
          }
        } catch (u) {
          i = { error: u };
        } finally {
          try {
            a && !a.done && (o = s.return) && o.call(s);
          } finally {
            if (i) throw i.error;
          }
        }
      }
    });
  }, t.prototype.error = function(n) {
    var r = this;
    Ie(function() {
      if (r._throwIfClosed(), !r.isStopped) {
        r.hasError = r.isStopped = !0, r.thrownError = n;
        for (var i = r.observers; i.length; )
          i.shift().error(n);
      }
    });
  }, t.prototype.complete = function() {
    var n = this;
    Ie(function() {
      if (n._throwIfClosed(), !n.isStopped) {
        n.isStopped = !0;
        for (var r = n.observers; r.length; )
          r.shift().complete();
      }
    });
  }, t.prototype.unsubscribe = function() {
    this.isStopped = this.closed = !0, this.observers = this.currentObservers = null;
  }, Object.defineProperty(t.prototype, "observed", {
    get: function() {
      var n;
      return ((n = this.observers) === null || n === void 0 ? void 0 : n.length) > 0;
    },
    enumerable: !1,
    configurable: !0
  }), t.prototype._trySubscribe = function(n) {
    return this._throwIfClosed(), e.prototype._trySubscribe.call(this, n);
  }, t.prototype._subscribe = function(n) {
    return this._throwIfClosed(), this._checkFinalizedStatuses(n), this._innerSubscribe(n);
  }, t.prototype._innerSubscribe = function(n) {
    var r = this, i = this, o = i.hasError, s = i.isStopped, a = i.observers;
    return o || s ? on : (this.currentObservers = null, a.push(n), new Ae(function() {
      r.currentObservers = null, $e(a, n);
    }));
  }, t.prototype._checkFinalizedStatuses = function(n) {
    var r = this, i = r.hasError, o = r.thrownError, s = r.isStopped;
    i ? n.error(o) : s && n.complete();
  }, t.prototype.asObservable = function() {
    var n = new k();
    return n.source = this, n;
  }, t.create = function(n, r) {
    return new kt(n, r);
  }, t;
}(k), kt = function(e) {
  ee(t, e);
  function t(n, r) {
    var i = e.call(this) || this;
    return i.destination = n, i.source = r, i;
  }
  return t.prototype.next = function(n) {
    var r, i;
    (i = (r = this.destination) === null || r === void 0 ? void 0 : r.next) === null || i === void 0 || i.call(r, n);
  }, t.prototype.error = function(n) {
    var r, i;
    (i = (r = this.destination) === null || r === void 0 ? void 0 : r.error) === null || i === void 0 || i.call(r, n);
  }, t.prototype.complete = function() {
    var n, r;
    (r = (n = this.destination) === null || n === void 0 ? void 0 : n.complete) === null || r === void 0 || r.call(n);
  }, t.prototype._subscribe = function(n) {
    var r, i;
    return (i = (r = this.source) === null || r === void 0 ? void 0 : r.subscribe(n)) !== null && i !== void 0 ? i : on;
  }, t;
}(J), cn = function(e) {
  ee(t, e);
  function t(n) {
    var r = e.call(this) || this;
    return r._value = n, r;
  }
  return Object.defineProperty(t.prototype, "value", {
    get: function() {
      return this.getValue();
    },
    enumerable: !1,
    configurable: !0
  }), t.prototype._subscribe = function(n) {
    var r = e.prototype._subscribe.call(this, n);
    return !r.closed && n.next(this._value), r;
  }, t.prototype.getValue = function() {
    var n = this, r = n.hasError, i = n.thrownError, o = n._value;
    if (r)
      throw i;
    return this._throwIfClosed(), o;
  }, t.prototype.next = function(n) {
    e.prototype.next.call(this, this._value = n);
  }, t;
}(J), Br = {
  now: function() {
    return Date.now();
  }
}, Vr = function(e) {
  ee(t, e);
  function t(n, r) {
    return e.call(this) || this;
  }
  return t.prototype.schedule = function(n, r) {
    return this;
  }, t;
}(Ae), Lt = {
  setInterval: function(e, t) {
    for (var n = [], r = 2; r < arguments.length; r++)
      n[r - 2] = arguments[r];
    return setInterval.apply(void 0, me([e, t], ae(n)));
  },
  clearInterval: function(e) {
    return clearInterval(e);
  },
  delegate: void 0
}, Hr = function(e) {
  ee(t, e);
  function t(n, r) {
    var i = e.call(this, n, r) || this;
    return i.scheduler = n, i.work = r, i.pending = !1, i;
  }
  return t.prototype.schedule = function(n, r) {
    var i;
    if (r === void 0 && (r = 0), this.closed)
      return this;
    this.state = n;
    var o = this.id, s = this.scheduler;
    return o != null && (this.id = this.recycleAsyncId(s, o, r)), this.pending = !0, this.delay = r, this.id = (i = this.id) !== null && i !== void 0 ? i : this.requestAsyncId(s, this.id, r), this;
  }, t.prototype.requestAsyncId = function(n, r, i) {
    return i === void 0 && (i = 0), Lt.setInterval(n.flush.bind(n, this), i);
  }, t.prototype.recycleAsyncId = function(n, r, i) {
    if (i === void 0 && (i = 0), i != null && this.delay === i && this.pending === !1)
      return r;
    r != null && Lt.clearInterval(r);
  }, t.prototype.execute = function(n, r) {
    if (this.closed)
      return new Error("executing a cancelled action");
    this.pending = !1;
    var i = this._execute(n, r);
    if (i)
      return i;
    this.pending === !1 && this.id != null && (this.id = this.recycleAsyncId(this.scheduler, this.id, null));
  }, t.prototype._execute = function(n, r) {
    var i = !1, o;
    try {
      this.work(n);
    } catch (s) {
      i = !0, o = s || new Error("Scheduled action threw falsy error");
    }
    if (i)
      return this.unsubscribe(), o;
  }, t.prototype.unsubscribe = function() {
    if (!this.closed) {
      var n = this, r = n.id, i = n.scheduler, o = i.actions;
      this.work = this.state = this.scheduler = null, this.pending = !1, $e(o, this), r != null && (this.id = this.recycleAsyncId(i, r, null)), this.delay = null, e.prototype.unsubscribe.call(this);
    }
  }, t;
}(Vr), _t = function() {
  function e(t, n) {
    n === void 0 && (n = e.now), this.schedulerActionCtor = t, this.now = n;
  }
  return e.prototype.schedule = function(t, n, r) {
    return n === void 0 && (n = 0), new this.schedulerActionCtor(this, t).schedule(r, n);
  }, e.now = Br.now, e;
}(), zr = function(e) {
  ee(t, e);
  function t(n, r) {
    r === void 0 && (r = _t.now);
    var i = e.call(this, n, r) || this;
    return i.actions = [], i._active = !1, i;
  }
  return t.prototype.flush = function(n) {
    var r = this.actions;
    if (this._active) {
      r.push(n);
      return;
    }
    var i;
    this._active = !0;
    do
      if (i = n.execute(n.state, n.delay))
        break;
    while (n = r.shift());
    if (this._active = !1, i) {
      for (; n = r.shift(); )
        n.unsubscribe();
      throw i;
    }
  }, t;
}(_t), Yr = new zr(Hr);
function Kr(e) {
  return e && C(e.schedule);
}
function Gr(e) {
  return e[e.length - 1];
}
function mt(e) {
  return Kr(Gr(e)) ? e.pop() : void 0;
}
var bt = function(e) {
  return e && typeof e.length == "number" && typeof e != "function";
};
function un(e) {
  return C(e == null ? void 0 : e.then);
}
function ln(e) {
  return C(e[ht]);
}
function fn(e) {
  return Symbol.asyncIterator && C(e == null ? void 0 : e[Symbol.asyncIterator]);
}
function dn(e) {
  return new TypeError("You provided " + (e !== null && typeof e == "object" ? "an invalid object" : "'" + e + "'") + " where a stream was expected. You can provide an Observable, Promise, ReadableStream, Array, AsyncIterable, or Iterable.");
}
function qr() {
  return typeof Symbol != "function" || !Symbol.iterator ? "@@iterator" : Symbol.iterator;
}
var pn = qr();
function hn(e) {
  return C(e == null ? void 0 : e[pn]);
}
function vn(e) {
  return kr(this, arguments, function() {
    var n, r, i, o;
    return rn(this, function(s) {
      switch (s.label) {
        case 0:
          n = e.getReader(), s.label = 1;
        case 1:
          s.trys.push([1, , 9, 10]), s.label = 2;
        case 2:
          return [4, de(n.read())];
        case 3:
          return r = s.sent(), i = r.value, o = r.done, o ? [4, de(void 0)] : [3, 5];
        case 4:
          return [2, s.sent()];
        case 5:
          return [4, de(i)];
        case 6:
          return [4, s.sent()];
        case 7:
          return s.sent(), [3, 2];
        case 8:
          return [3, 10];
        case 9:
          return n.releaseLock(), [7];
        case 10:
          return [2];
      }
    });
  });
}
function mn(e) {
  return C(e == null ? void 0 : e.getReader);
}
function H(e) {
  if (e instanceof k)
    return e;
  if (e != null) {
    if (ln(e))
      return Qr(e);
    if (bt(e))
      return Jr(e);
    if (un(e))
      return Xr(e);
    if (fn(e))
      return bn(e);
    if (hn(e))
      return Zr(e);
    if (mn(e))
      return ei(e);
  }
  throw dn(e);
}
function Qr(e) {
  return new k(function(t) {
    var n = e[ht]();
    if (C(n.subscribe))
      return n.subscribe(t);
    throw new TypeError("Provided object does not correctly implement Symbol.observable");
  });
}
function Jr(e) {
  return new k(function(t) {
    for (var n = 0; n < e.length && !t.closed; n++)
      t.next(e[n]);
    t.complete();
  });
}
function Xr(e) {
  return new k(function(t) {
    e.then(function(n) {
      t.closed || (t.next(n), t.complete());
    }, function(n) {
      return t.error(n);
    }).then(null, an);
  });
}
function Zr(e) {
  return new k(function(t) {
    var n, r;
    try {
      for (var i = ve(e), o = i.next(); !o.done; o = i.next()) {
        var s = o.value;
        if (t.next(s), t.closed)
          return;
      }
    } catch (a) {
      n = { error: a };
    } finally {
      try {
        o && !o.done && (r = i.return) && r.call(i);
      } finally {
        if (n) throw n.error;
      }
    }
    t.complete();
  });
}
function bn(e) {
  return new k(function(t) {
    ti(e, t).catch(function(n) {
      return t.error(n);
    });
  });
}
function ei(e) {
  return bn(vn(e));
}
function ti(e, t) {
  var n, r, i, o;
  return Or(this, void 0, void 0, function() {
    var s, a;
    return rn(this, function(c) {
      switch (c.label) {
        case 0:
          c.trys.push([0, 5, 6, 11]), n = Lr(e), c.label = 1;
        case 1:
          return [4, n.next()];
        case 2:
          if (r = c.sent(), !!r.done) return [3, 4];
          if (s = r.value, t.next(s), t.closed)
            return [2];
          c.label = 3;
        case 3:
          return [3, 1];
        case 4:
          return [3, 11];
        case 5:
          return a = c.sent(), i = { error: a }, [3, 11];
        case 6:
          return c.trys.push([6, , 9, 10]), r && !r.done && (o = n.return) ? [4, o.call(n)] : [3, 8];
        case 7:
          c.sent(), c.label = 8;
        case 8:
          return [3, 10];
        case 9:
          if (i) throw i.error;
          return [7];
        case 10:
          return [7];
        case 11:
          return t.complete(), [2];
      }
    });
  });
}
function re(e, t, n, r, i) {
  r === void 0 && (r = 0), i === void 0 && (i = !1);
  var o = t.schedule(function() {
    n(), i ? e.add(this.schedule(null, r)) : this.unsubscribe();
  }, r);
  if (e.add(o), !i)
    return o;
}
function gn(e, t) {
  return t === void 0 && (t = 0), F(function(n, r) {
    n.subscribe(B(r, function(i) {
      return re(r, e, function() {
        return r.next(i);
      }, t);
    }, function() {
      return re(r, e, function() {
        return r.complete();
      }, t);
    }, function(i) {
      return re(r, e, function() {
        return r.error(i);
      }, t);
    }));
  });
}
function yn(e, t) {
  return t === void 0 && (t = 0), F(function(n, r) {
    r.add(e.schedule(function() {
      return n.subscribe(r);
    }, t));
  });
}
function ni(e, t) {
  return H(e).pipe(yn(t), gn(t));
}
function ri(e, t) {
  return H(e).pipe(yn(t), gn(t));
}
function ii(e, t) {
  return new k(function(n) {
    var r = 0;
    return t.schedule(function() {
      r === e.length ? n.complete() : (n.next(e[r++]), n.closed || this.schedule());
    });
  });
}
function oi(e, t) {
  return new k(function(n) {
    var r;
    return re(n, t, function() {
      r = e[pn](), re(n, t, function() {
        var i, o, s;
        try {
          i = r.next(), o = i.value, s = i.done;
        } catch (a) {
          n.error(a);
          return;
        }
        s ? n.complete() : n.next(o);
      }, 0, !0);
    }), function() {
      return C(r == null ? void 0 : r.return) && r.return();
    };
  });
}
function wn(e, t) {
  if (!e)
    throw new Error("Iterable cannot be null");
  return new k(function(n) {
    re(n, t, function() {
      var r = e[Symbol.asyncIterator]();
      re(n, t, function() {
        r.next().then(function(i) {
          i.done ? n.complete() : n.next(i.value);
        });
      }, 0, !0);
    });
  });
}
function si(e, t) {
  return wn(vn(e), t);
}
function ai(e, t) {
  if (e != null) {
    if (ln(e))
      return ni(e, t);
    if (bt(e))
      return ii(e, t);
    if (un(e))
      return ri(e, t);
    if (fn(e))
      return wn(e, t);
    if (hn(e))
      return oi(e, t);
    if (mn(e))
      return si(e, t);
  }
  throw dn(e);
}
function Ye(e, t) {
  return t ? ai(e, t) : H(e);
}
function It() {
  for (var e = [], t = 0; t < arguments.length; t++)
    e[t] = arguments[t];
  var n = mt(e);
  return Ye(e, n);
}
function ci(e) {
  return e instanceof Date && !isNaN(e);
}
var ui = dt(function(e) {
  return function(n) {
    n === void 0 && (n = null), e(this), this.message = "Timeout has occurred", this.name = "TimeoutError", this.info = n;
  };
});
function li(e, t) {
  var n = ci(e) ? { first: e } : typeof e == "number" ? { each: e } : e, r = n.first, i = n.each, o = n.with, s = o === void 0 ? fi : o, a = n.scheduler, c = a === void 0 ? Yr : a, u = n.meta, l = u === void 0 ? null : u;
  if (r == null && i == null)
    throw new TypeError("No timeout provided.");
  return F(function(f, h) {
    var d, m, p = null, b = 0, v = function(S) {
      m = re(h, c, function() {
        try {
          d.unsubscribe(), H(s({
            meta: l,
            lastValue: p,
            seen: b
          })).subscribe(h);
        } catch (E) {
          h.error(E);
        }
      }, S);
    };
    d = f.subscribe(B(h, function(S) {
      m == null || m.unsubscribe(), b++, h.next(p = S), i > 0 && v(i);
    }, void 0, void 0, function() {
      m != null && m.closed || m == null || m.unsubscribe(), p = null;
    })), !b && v(r != null ? typeof r == "number" ? r : +r - c.now() : i);
  });
}
function fi(e) {
  throw new ui(e);
}
function P(e, t) {
  return F(function(n, r) {
    var i = 0;
    n.subscribe(B(r, function(o) {
      r.next(e.call(t, o, i++));
    }));
  });
}
var di = Array.isArray;
function pi(e, t) {
  return di(t) ? e.apply(void 0, me([], ae(t))) : e(t);
}
function hi(e) {
  return P(function(t) {
    return pi(e, t);
  });
}
function vi(e, t, n, r, i, o, s, a) {
  var c = [], u = 0, l = 0, f = !1, h = function() {
    f && !c.length && !u && t.complete();
  }, d = function(p) {
    return u < r ? m(p) : c.push(p);
  }, m = function(p) {
    u++;
    var b = !1;
    H(n(p, l++)).subscribe(B(t, function(v) {
      t.next(v);
    }, function() {
      b = !0;
    }, void 0, function() {
      if (b)
        try {
          u--;
          for (var v = function() {
            var S = c.shift();
            s || m(S);
          }; c.length && u < r; )
            v();
          h();
        } catch (S) {
          t.error(S);
        }
    }));
  };
  return e.subscribe(B(t, d, function() {
    f = !0, h();
  })), function() {
  };
}
function gt(e, t, n) {
  return n === void 0 && (n = 1 / 0), C(t) ? gt(function(r, i) {
    return P(function(o, s) {
      return t(r, o, i, s);
    })(H(e(r, i)));
  }, n) : (typeof t == "number" && (n = t), F(function(r, i) {
    return vi(r, i, e, n);
  }));
}
function mi(e) {
  return gt(vt, e);
}
function bi() {
  return mi(1);
}
function Ue() {
  for (var e = [], t = 0; t < arguments.length; t++)
    e[t] = arguments[t];
  return bi()(Ye(e, mt(e)));
}
var gi = ["addListener", "removeListener"], yi = ["addEventListener", "removeEventListener"], wi = ["on", "off"];
function ot(e, t, n, r) {
  if (C(n) && (r = n, n = void 0), r)
    return ot(e, t, n).pipe(hi(r));
  var i = ae(Ti(e) ? yi.map(function(a) {
    return function(c) {
      return e[a](t, c, n);
    };
  }) : xi(e) ? gi.map(Mt(e, t)) : Si(e) ? wi.map(Mt(e, t)) : [], 2), o = i[0], s = i[1];
  if (!o && bt(e))
    return gt(function(a) {
      return ot(a, t, n);
    })(H(e));
  if (!o)
    throw new TypeError("Invalid event target");
  return new k(function(a) {
    var c = function() {
      for (var u = [], l = 0; l < arguments.length; l++)
        u[l] = arguments[l];
      return a.next(1 < u.length ? u : u[0]);
    };
    return o(c), function() {
      return s(c);
    };
  });
}
function Mt(e, t) {
  return function(n) {
    return function(r) {
      return e[n](t, r);
    };
  };
}
function xi(e) {
  return C(e.addListener) && C(e.removeListener);
}
function Si(e) {
  return C(e.on) && C(e.off);
}
function Ti(e) {
  return C(e.addEventListener) && C(e.removeEventListener);
}
function Ke(e, t) {
  return F(function(n, r) {
    var i = 0;
    n.subscribe(B(r, function(o) {
      return e.call(t, o, i++) && r.next(o);
    }));
  });
}
function Ei(e, t, n, r, i) {
  return function(o, s) {
    var a = n, c = t, u = 0;
    o.subscribe(B(s, function(l) {
      var f = u++;
      c = a ? e(c, l, f) : (a = !0, l), s.next(c);
    }, i));
  };
}
function Ci(e, t) {
  return t === void 0 && (t = vt), e = e ?? Pi, F(function(n, r) {
    var i, o = !0;
    n.subscribe(B(r, function(s) {
      var a = t(s);
      (o || !e(i, a)) && (o = !1, i = a, r.next(s));
    }));
  });
}
function Pi(e, t) {
  return e === t;
}
function Ai(e) {
  return F(function(t, n) {
    try {
      t.subscribe(n);
    } finally {
      n.add(e);
    }
  });
}
function st(e, t) {
  return F(Ei(e, t, arguments.length >= 2, !0));
}
function Oi(e) {
  e === void 0 && (e = {});
  var t = e.connector, n = t === void 0 ? function() {
    return new J();
  } : t, r = e.resetOnError, i = r === void 0 ? !0 : r, o = e.resetOnComplete, s = o === void 0 ? !0 : o, a = e.resetOnRefCountZero, c = a === void 0 ? !0 : a;
  return function(u) {
    var l, f, h, d = 0, m = !1, p = !1, b = function() {
      f == null || f.unsubscribe(), f = void 0;
    }, v = function() {
      b(), l = h = void 0, m = p = !1;
    }, S = function() {
      var E = l;
      v(), E == null || E.unsubscribe();
    };
    return F(function(E, A) {
      d++, !p && !m && b();
      var y = h = h ?? n();
      A.add(function() {
        d--, d === 0 && !p && !m && (f = Je(S, c));
      }), y.subscribe(A), !l && d > 0 && (l = new Ee({
        next: function(x) {
          return y.next(x);
        },
        error: function(x) {
          p = !0, b(), f = Je(v, i, x), y.error(x);
        },
        complete: function() {
          m = !0, b(), f = Je(v, s), y.complete();
        }
      }), H(E).subscribe(l));
    })(u);
  };
}
function Je(e, t) {
  for (var n = [], r = 2; r < arguments.length; r++)
    n[r - 2] = arguments[r];
  if (t === !0) {
    e();
    return;
  }
  if (t !== !1) {
    var i = new Ee({
      next: function() {
        i.unsubscribe(), e();
      }
    });
    return H(t.apply(void 0, me([], ae(n)))).subscribe(i);
  }
}
function ki(e) {
  return F(function(t, n) {
    var r = !1, i = B(n, function() {
      i == null || i.unsubscribe(), r = !0;
    }, Ne);
    H(e).subscribe(i), t.subscribe(B(n, function(o) {
      return r && n.next(o);
    }));
  });
}
function R() {
  for (var e = [], t = 0; t < arguments.length; t++)
    e[t] = arguments[t];
  var n = mt(e);
  return F(function(r, i) {
    (n ? Ue(e, r, n) : Ue(e, r)).subscribe(i);
  });
}
function xn(e, t) {
  return F(function(n, r) {
    var i = null, o = 0, s = !1, a = function() {
      return s && !i && r.complete();
    };
    n.subscribe(B(r, function(c) {
      i == null || i.unsubscribe();
      var u = 0, l = o++;
      H(e(c, l)).subscribe(i = B(r, function(f) {
        return r.next(t ? t(c, f, l, u++) : f);
      }, function() {
        i = null, a();
      }));
    }, function() {
      s = !0, a();
    }));
  });
}
function Dt(e) {
  return F(function(t, n) {
    H(e).subscribe(B(n, function() {
      return n.complete();
    }, Ne)), !n.closed && t.subscribe(n);
  });
}
var Li = Object.defineProperty, _i = Object.defineProperties, Ii = Object.getOwnPropertyDescriptors, Rt = Object.getOwnPropertySymbols, Mi = Object.prototype.hasOwnProperty, Di = Object.prototype.propertyIsEnumerable, $t = (e, t, n) => t in e ? Li(e, t, { enumerable: !0, configurable: !0, writable: !0, value: n }) : e[t] = n, X = (e, t) => {
  for (var n in t || (t = {}))
    Mi.call(t, n) && $t(e, n, t[n]);
  if (Rt)
    for (var n of Rt(t))
      Di.call(t, n) && $t(e, n, t[n]);
  return e;
}, Se = (e, t) => _i(e, Ii(t)), V = (e, t, n) => new Promise((r, i) => {
  var o = (c) => {
    try {
      a(n.next(c));
    } catch (u) {
      i(u);
    }
  }, s = (c) => {
    try {
      a(n.throw(c));
    } catch (u) {
      i(u);
    }
  }, a = (c) => c.done ? r(c.value) : Promise.resolve(c.value).then(o, s);
  a((n = n.apply(e, t)).next());
}), Sn = "lk";
function W(e) {
  return typeof e > "u" ? !1 : Ri(e) || $i(e);
}
function Ri(e) {
  var t;
  return e ? e.hasOwnProperty("participant") && e.hasOwnProperty("source") && e.hasOwnProperty("track") && typeof ((t = e.publication) == null ? void 0 : t.track) < "u" : !1;
}
function $i(e) {
  return e ? e.hasOwnProperty("participant") && e.hasOwnProperty("source") && e.hasOwnProperty("publication") && typeof e.publication < "u" : !1;
}
function Ce(e) {
  return e ? e.hasOwnProperty("participant") && e.hasOwnProperty("source") && typeof e.publication > "u" : !1;
}
function N(e) {
  if (typeof e == "string" || typeof e == "number")
    return `${e}`;
  if (Ce(e))
    return `${e.participant.identity}_${e.source}_placeholder`;
  if (W(e))
    return `${e.participant.identity}_${e.publication.source}_${e.publication.trackSid}`;
  throw new Error(`Can't generate a id for the given track reference: ${e}`);
}
function Io(e, t) {
  return e === void 0 || t === void 0 ? !1 : W(e) && W(t) ? e.publication.trackSid === t.publication.trackSid : N(e) === N(t);
}
function Mo(e, t) {
  return typeof t > "u" ? !1 : W(e) ? t.some(
    (n) => n.participant.identity === e.participant.identity && W(n) && n.publication.trackSid === e.publication.trackSid
  ) : Ce(e) ? t.some(
    (n) => n.participant.identity === e.participant.identity && Ce(n) && n.source === e.source
  ) : !1;
}
function Ni(e, t) {
  return Ce(e) && W(t) && t.participant.identity === e.participant.identity && t.source === e.source;
}
function Do() {
  const e = document.createElement("p");
  e.style.width = "100%", e.style.height = "200px";
  const t = document.createElement("div");
  t.style.position = "absolute", t.style.top = "0px", t.style.left = "0px", t.style.visibility = "hidden", t.style.width = "200px", t.style.height = "150px", t.style.overflow = "hidden", t.appendChild(e), document.body.appendChild(t);
  const n = e.offsetWidth;
  t.style.overflow = "scroll";
  let r = e.offsetWidth;
  return n === r && (r = t.clientWidth), document.body.removeChild(t), n - r;
}
function Ro() {
  return typeof document < "u";
}
function Ui(e) {
  e = X({}, e);
  const t = "(?:(?:[a-z]+:)?//)?", n = "(?:\\S+(?::\\S*)?@)?", r = new RegExp(
    "(?:25[0-5]|2[0-4]\\d|1\\d\\d|[1-9]\\d|\\d)(?:\\.(?:25[0-5]|2[0-4]\\d|1\\d\\d|[1-9]\\d|\\d)){3}",
    "g"
  ).source, u = `(?:${t}|www\\.)${n}(?:localhost|${r}|(?:(?:[a-z\\u00a1-\\uffff0-9][-_]*)*[a-z\\u00a1-\\uffff0-9]+)(?:\\.(?:[a-z\\u00a1-\\uffff0-9]-*)*[a-z\\u00a1-\\uffff0-9]+)*(?:\\.(?:[a-z\\u00a1-\\uffff]{2,}))\\.?)(?::\\d{2,5})?(?:[/?#][^\\s"]*)?`;
  return e.exact ? new RegExp(`(?:^${u}$)`, "i") : new RegExp(u, "ig");
}
var Nt = "[^\\.\\s@:](?:[^\\s@:]*[^\\s@:\\.])?@[^\\.\\s@]+(?:\\.[^\\.\\s@]+)*";
function Fi({ exact: e } = {}) {
  return e ? new RegExp(`^${Nt}$`) : new RegExp(Nt, "g");
}
function $o(e, t) {
  return V(this, null, function* () {
    const { x: n, y: r } = yield Sr(e, t, {
      placement: "top",
      middleware: [yr(6), xr(), wr({ padding: 5 })]
    });
    return { x: n, y: r };
  });
}
function No(e, t) {
  return !e.contains(t.target);
}
var Uo = () => ({
  email: Fi(),
  url: Ui({})
});
function Fo(e, t) {
  const n = Object.entries(t).map(
    ([o, s], a) => Array.from(e.matchAll(s)).map(({ index: c, 0: u }) => ({
      type: o,
      weight: a,
      content: u,
      index: c ?? 0
    }))
  ).flat().sort((o, s) => {
    const a = o.index - s.index;
    return a !== 0 ? a : o.weight - s.weight;
  }).filter(({ index: o }, s, a) => {
    if (s === 0) return !0;
    const c = a[s - 1];
    return c.index + c.content.length <= o;
  }), r = [];
  let i = 0;
  for (const { type: o, content: s, index: a } of n)
    a > i && r.push(e.substring(i, a)), r.push({ type: o, content: s }), i = a + s.length;
  return e.length > i && r.push(e.substring(i)), r;
}
var ji = [
  g.ConnectionStateChanged,
  g.RoomMetadataChanged,
  g.ActiveSpeakersChanged,
  g.ConnectionQualityChanged,
  g.ParticipantConnected,
  g.ParticipantDisconnected,
  g.ParticipantPermissionsChanged,
  g.ParticipantMetadataChanged,
  g.ParticipantNameChanged,
  g.ParticipantAttributesChanged,
  g.TrackMuted,
  g.TrackUnmuted,
  g.TrackPublished,
  g.TrackUnpublished,
  g.TrackStreamStateChanged,
  g.TrackSubscriptionFailed,
  g.TrackSubscriptionPermissionChanged,
  g.TrackSubscriptionStatusChanged
], Tn = [
  ...ji,
  g.LocalTrackPublished,
  g.LocalTrackUnpublished
], Wi = [
  w.TrackPublished,
  w.TrackUnpublished,
  w.TrackMuted,
  w.TrackUnmuted,
  w.TrackStreamStateChanged,
  w.TrackSubscribed,
  w.TrackUnsubscribed,
  w.TrackSubscriptionPermissionChanged,
  w.TrackSubscriptionFailed,
  w.LocalTrackPublished,
  w.LocalTrackUnpublished
], Bi = [
  w.ConnectionQualityChanged,
  w.IsSpeakingChanged,
  w.ParticipantMetadataChanged,
  w.ParticipantPermissionsChanged,
  w.TrackMuted,
  w.TrackUnmuted,
  w.TrackPublished,
  w.TrackUnpublished,
  w.TrackStreamStateChanged,
  w.TrackSubscriptionFailed,
  w.TrackSubscriptionPermissionChanged,
  w.TrackSubscriptionStatusChanged
], En = [
  ...Bi,
  w.LocalTrackPublished,
  w.LocalTrackUnpublished
], _ = Ar.getLogger("lk-components-js");
_.setDefaultLevel("WARN");
function jo(e, t = {}) {
  var n;
  _.setLevel(e), Wn((n = t.liveKitClientLogLevel) != null ? n : e);
}
function Wo(e, t = {}) {
  var n;
  const r = _.methodFactory;
  _.methodFactory = (i, o, s) => {
    const a = r(i, o, s), c = yt[i], u = c >= o && c < yt.silent;
    return (l, f) => {
      f ? a(l, f) : a(l), u && e(c, l, f);
    };
  }, _.setLevel(_.getLevel()), Bn((n = t.liveKitClientLogExtension) != null ? n : e);
}
var Bo = [
  {
    columns: 1,
    rows: 1
  },
  {
    columns: 1,
    rows: 2,
    orientation: "portrait"
  },
  {
    columns: 2,
    rows: 1,
    orientation: "landscape"
  },
  {
    columns: 2,
    rows: 2,
    minWidth: 560
  },
  {
    columns: 3,
    rows: 3,
    minWidth: 700
  },
  {
    columns: 4,
    rows: 4,
    minWidth: 960
  },
  {
    columns: 5,
    rows: 5,
    minWidth: 1100
  }
];
function Vi(e, t, n, r) {
  if (e.length < 1)
    throw new Error("At least one grid layout definition must be provided.");
  const i = Hi(e);
  if (n <= 0 || r <= 0)
    return i[0];
  let o = 0;
  const s = n / r > 1 ? "landscape" : "portrait";
  let a = i.find((c, u, l) => {
    o = u;
    const f = l.findIndex((h, d) => {
      const m = !h.orientation || h.orientation === s, p = d > u, b = h.maxTiles === c.maxTiles;
      return p && b && m;
    }) !== -1;
    return c.maxTiles >= t && !f;
  });
  if (a === void 0)
    if (a = i[i.length - 1], a)
      _.warn(
        `No layout found for: participantCount: ${t}, width/height: ${n}/${r} fallback to biggest available layout (${a}).`
      );
    else
      throw new Error("No layout or fallback layout found.");
  if ((n < a.minWidth || r < a.minHeight) && o > 0) {
    const c = i[o - 1];
    a = Vi(
      i.slice(0, o),
      c.maxTiles,
      n,
      r
    );
  }
  return a;
}
function Hi(e) {
  return [...e].map((t) => {
    var n, r;
    return {
      name: `${t.columns}x${t.rows}`,
      columns: t.columns,
      rows: t.rows,
      maxTiles: t.columns * t.rows,
      minWidth: (n = t.minWidth) != null ? n : 0,
      minHeight: (r = t.minHeight) != null ? r : 0,
      orientation: t.orientation
    };
  }).sort((t, n) => t.maxTiles !== n.maxTiles ? t.maxTiles - n.maxTiles : t.minWidth !== 0 || n.minWidth !== 0 ? t.minWidth - n.minWidth : t.minHeight !== 0 || n.minHeight !== 0 ? t.minHeight - n.minHeight : 0);
}
function Vo() {
  return typeof navigator < "u" && navigator.mediaDevices && !!navigator.mediaDevices.getDisplayMedia;
}
function Ho(e, t) {
  var n;
  return Se(X({}, e), {
    receivedAtMediaTimestamp: (n = t.rtpTimestamp) != null ? n : 0,
    receivedAt: t.timestamp
  });
}
function zo(e, t, n) {
  return [...e, ...t].reduceRight((r, i) => (r.find((o) => o.id === i.id) || r.unshift(i), r), []).slice(0 - n);
}
var Cn = [], Pn = {
  showChat: !1,
  unreadMessages: 0,
  showSettings: !1
};
function zi(e) {
  return typeof e == "object";
}
function Yo(e) {
  return Array.isArray(e) && e.filter(zi).length > 0;
}
function An(e, t) {
  return t.audioLevel - e.audioLevel;
}
function On(e, t) {
  return e.isSpeaking === t.isSpeaking ? 0 : e.isSpeaking ? -1 : 1;
}
function kn(e, t) {
  var n, r, i, o;
  return e.lastSpokeAt !== void 0 || t.lastSpokeAt !== void 0 ? ((r = (n = t.lastSpokeAt) == null ? void 0 : n.getTime()) != null ? r : 0) - ((o = (i = e.lastSpokeAt) == null ? void 0 : i.getTime()) != null ? o : 0) : 0;
}
function Fe(e, t) {
  var n, r, i, o;
  return ((r = (n = e.joinedAt) == null ? void 0 : n.getTime()) != null ? r : 0) - ((o = (i = t.joinedAt) == null ? void 0 : i.getTime()) != null ? o : 0);
}
function Yi(e, t) {
  return W(e) ? W(t) ? 0 : -1 : W(t) ? 1 : 0;
}
function Ki(e, t) {
  const n = e.participant.isCameraEnabled, r = t.participant.isCameraEnabled;
  return n !== r ? n ? -1 : 1 : 0;
}
function Ko(e) {
  const t = [], n = [], r = [], i = [];
  e.forEach((a) => {
    a.participant.isLocal && a.source === D.Source.Camera ? t.push(a) : a.source === D.Source.ScreenShare ? n.push(a) : a.source === D.Source.Camera ? r.push(a) : i.push(a);
  });
  const o = Gi(n), s = qi(r);
  return [...t, ...o, ...s, ...i];
}
function Gi(e) {
  const t = [], n = [];
  return e.forEach((i) => {
    i.participant.isLocal ? t.push(i) : n.push(i);
  }), t.sort((i, o) => Fe(i.participant, o.participant)), n.sort((i, o) => Fe(i.participant, o.participant)), [...n, ...t];
}
function qi(e) {
  const t = [], n = [];
  return e.forEach((r) => {
    r.participant.isLocal ? t.push(r) : n.push(r);
  }), n.sort((r, i) => r.participant.isSpeaking && i.participant.isSpeaking ? An(r.participant, i.participant) : r.participant.isSpeaking !== i.participant.isSpeaking ? On(r.participant, i.participant) : r.participant.lastSpokeAt !== i.participant.lastSpokeAt ? kn(r.participant, i.participant) : W(r) !== W(i) ? Yi(r, i) : r.participant.isCameraEnabled !== i.participant.isCameraEnabled ? Ki(r, i) : Fe(r.participant, i.participant)), [...t, ...n];
}
function Go(e) {
  const t = [...e];
  t.sort((r, i) => {
    if (r.isSpeaking && i.isSpeaking)
      return An(r, i);
    if (r.isSpeaking !== i.isSpeaking)
      return On(r, i);
    if (r.lastSpokeAt !== i.lastSpokeAt)
      return kn(r, i);
    const o = r.videoTrackPublications.size > 0, s = i.videoTrackPublications.size > 0;
    return o !== s ? o ? -1 : 1 : Fe(r, i);
  });
  const n = t.find((r) => r.isLocal);
  if (n) {
    const r = t.indexOf(n);
    r >= 0 && (t.splice(r, 1), t.length > 0 ? t.splice(0, 0, n) : t.push(n));
  }
  return t;
}
function Qi(e, t) {
  return e.reduce(
    (n, r, i) => i % t === 0 ? [...n, [r]] : [...n.slice(0, -1), [...n.slice(-1)[0], r]],
    []
  );
}
function Ut(e, t) {
  const n = Math.max(e.length, t.length);
  return new Array(n).fill([]).map((r, i) => [e[i], t[i]]);
}
function je(e, t, n) {
  return e.filter((r) => !t.map((i) => n(i)).includes(n(r)));
}
function at(e) {
  return e.map((t) => typeof t == "string" || typeof t == "number" ? `${t}` : N(t));
}
function Ji(e, t) {
  return {
    dropped: je(e, t, N),
    added: je(t, e, N)
  };
}
function Xi(e) {
  return e.added.length !== 0 || e.dropped.length !== 0;
}
function ct(e, t) {
  const n = t.findIndex(
    (r) => N(r) === N(e)
  );
  if (n === -1)
    throw new Error(
      `Element not part of the array: ${N(
        e
      )} not in ${at(t)}`
    );
  return n;
}
function Zi(e, t, n) {
  const r = ct(e, n), i = ct(t, n);
  return n.splice(r, 1, t), n.splice(i, 1, e), n;
}
function eo(e, t) {
  const n = ct(e, t);
  return t.splice(n, 1), t;
}
function to(e, t) {
  return [...t, e];
}
function Xe(e, t) {
  return Qi(e, t);
}
function qo(e, t, n) {
  let r = no(e, t);
  if (r.length < t.length) {
    const s = je(t, r, N);
    r = [...r, ...s];
  }
  const i = Xe(r, n), o = Xe(t, n);
  if (Ut(i, o).forEach(([s, a], c) => {
    if (s && a) {
      const u = Xe(r, n)[c], l = Ji(u, a);
      Xi(l) && (_.debug(
        `Detected visual changes on page: ${c}, current: ${at(
          s
        )}, next: ${at(a)}`,
        { changes: l }
      ), l.added.length === l.dropped.length && Ut(l.added, l.dropped).forEach(([f, h]) => {
        if (f && h)
          r = Zi(f, h, r);
        else
          throw new Error(
            `For a swap action we need a addition and a removal one is missing: ${f}, ${h}`
          );
      }), l.added.length === 0 && l.dropped.length > 0 && l.dropped.forEach((f) => {
        r = eo(f, r);
      }), l.added.length > 0 && l.dropped.length === 0 && l.added.forEach((f) => {
        r = to(f, r);
      }));
    }
  }), r.length > t.length) {
    const s = je(r, t, N);
    r = r.filter(
      (a) => !s.map(N).includes(N(a))
    );
  }
  return r;
}
function no(e, t) {
  return e.map((n) => {
    const r = t.find(
      (i) => (
        // If the IDs match or ..
        N(n) === N(i) || // ... if the current item is a placeholder and the new item is the track reference can replace it.
        typeof n != "number" && Ce(n) && W(i) && Ni(n, i)
      )
    );
    return r ?? n;
  });
}
function U(e) {
  return `${Sn}-${e}`;
}
function Qo(e) {
  const t = Ft(e), n = Ln(e.participant).pipe(
    P(() => Ft(e)),
    R(t)
  );
  return { className: U(
    e.source === D.Source.Camera || e.source === D.Source.ScreenShare ? "participant-media-video" : "participant-media-audio"
  ), trackObserver: n };
}
function Ft(e) {
  if (W(e))
    return e.publication;
  {
    const { source: t, name: n, participant: r } = e;
    if (t && n)
      return r.getTrackPublications().find((i) => i.source === t && i.trackName === n);
    if (n)
      return r.getTrackPublicationByName(n);
    if (t)
      return r.getTrackPublication(t);
    throw new Error("At least one of source and name needs to be defined");
  }
}
function ce(e, ...t) {
  return new k((r) => {
    const i = () => {
      r.next(e);
    };
    return t.forEach((s) => {
      e.on(s, i);
    }), () => {
      t.forEach((s) => {
        e.off(s, i);
      });
    };
  }).pipe(R(e));
}
function ge(e, t) {
  return new k((r) => {
    const i = (...s) => {
      r.next(s);
    };
    return e.on(t, i), () => {
      e.off(t, i);
    };
  });
}
function Jo(e) {
  return ge(e, g.ConnectionStateChanged).pipe(
    P(([t]) => t),
    R(e.state)
  );
}
function Xo(e) {
  return ce(
    e,
    g.RoomMetadataChanged,
    g.ConnectionStateChanged
  ).pipe(
    P((n) => ({ name: n.name, metadata: n.metadata }))
  );
}
function Zo(e) {
  return ge(e, g.ActiveSpeakersChanged).pipe(
    P(([t]) => t)
  );
}
function es(e, t, n = !0) {
  var r;
  const i = () => V(this, null, function* () {
    try {
      const a = yield wt.getLocalDevices(e, n);
      o.next(a);
    } catch (a) {
      t == null || t(a);
    }
  }), o = new J(), s = o.pipe(
    Ai(() => {
      var a;
      (a = navigator == null ? void 0 : navigator.mediaDevices) == null || a.removeEventListener("devicechange", i);
    })
  );
  if (typeof window < "u") {
    if (!window.isSecureContext)
      throw new Error(
        "Accessing media devices is available only in secure contexts (HTTPS and localhost), in some or all supporting browsers. See: https://developer.mozilla.org/en-US/docs/Web/API/Navigator/mediaDevices"
      );
    (r = navigator == null ? void 0 : navigator.mediaDevices) == null || r.addEventListener("devicechange", i);
  }
  return Ue(
    wt.getLocalDevices(e, n).catch((a) => (t == null || t(a), [])),
    s
  );
}
function ro(e) {
  return ge(e, g.DataReceived);
}
function io(e) {
  return ce(e, g.AudioPlaybackStatusChanged).pipe(
    P((n) => ({ canPlayAudio: n.canPlaybackAudio }))
  );
}
function oo(e) {
  return ce(e, g.VideoPlaybackStatusChanged).pipe(
    P((n) => ({ canPlayVideo: n.canPlaybackVideo }))
  );
}
function so(e, t) {
  return ge(e, g.ActiveDeviceChanged).pipe(
    Ke(([n]) => n === t),
    P(([n, r]) => (_.debug("activeDeviceObservable | RoomEvent.ActiveDeviceChanged", { kind: n, deviceId: r }), r))
  );
}
function ts(e, t) {
  return ge(e, g.ParticipantEncryptionStatusChanged).pipe(
    Ke(
      ([, n]) => (t == null ? void 0 : t.identity) === (n == null ? void 0 : n.identity) || !n && (t == null ? void 0 : t.identity) === e.localParticipant.identity
    ),
    P(([n]) => n),
    R(
      t != null && t.isLocal ? t.isE2EEEnabled : !!(t != null && t.isEncrypted)
    )
  );
}
function ns(e) {
  return ge(e, g.RecordingStatusChanged).pipe(
    P(([t]) => t),
    R(e.isRecording)
  );
}
function ye(e, ...t) {
  return new k((r) => {
    const i = () => {
      r.next(e);
    };
    return t.forEach((s) => {
      e.on(s, i);
    }), () => {
      t.forEach((s) => {
        e.off(s, i);
      });
    };
  }).pipe(R(e));
}
function Ln(e) {
  return ye(
    e,
    w.TrackMuted,
    w.TrackUnmuted,
    w.ParticipantPermissionsChanged,
    // ParticipantEvent.IsSpeakingChanged,
    w.TrackPublished,
    w.TrackUnpublished,
    w.LocalTrackPublished,
    w.LocalTrackUnpublished,
    w.MediaDevicesError,
    w.TrackSubscriptionStatusChanged
    // ParticipantEvent.ConnectionQualityChanged,
  ).pipe(
    P((n) => {
      const { isMicrophoneEnabled: r, isCameraEnabled: i, isScreenShareEnabled: o } = n, s = n.getTrackPublication(D.Source.Microphone), a = n.getTrackPublication(D.Source.Camera);
      return {
        isCameraEnabled: i,
        isMicrophoneEnabled: r,
        isScreenShareEnabled: o,
        cameraTrack: a,
        microphoneTrack: s,
        participant: n
      };
    })
  );
}
function ao(e) {
  return e ? ye(
    e,
    w.ParticipantMetadataChanged,
    w.ParticipantNameChanged
  ).pipe(
    P(({ name: n, identity: r, metadata: i }) => ({
      name: n,
      identity: r,
      metadata: i
    })),
    R({
      name: e.name,
      identity: e.identity,
      metadata: e.metadata
    })
  ) : void 0;
}
function co(e) {
  return Ge(
    e,
    w.ConnectionQualityChanged
  ).pipe(
    P(([n]) => n),
    R(e.connectionQuality)
  );
}
function Ge(e, t) {
  return new k((r) => {
    const i = (...s) => {
      r.next(s);
    };
    return e.on(t, i), () => {
      e.off(t, i);
    };
  });
}
function uo(e) {
  var t, n, r, i;
  return ye(
    e.participant,
    w.TrackMuted,
    w.TrackUnmuted,
    w.TrackSubscribed,
    w.TrackUnsubscribed,
    w.LocalTrackPublished,
    w.LocalTrackUnpublished
  ).pipe(
    P((o) => {
      var s, a;
      const c = (s = e.publication) != null ? s : o.getTrackPublication(e.source);
      return (a = c == null ? void 0 : c.isMuted) != null ? a : !0;
    }),
    R(
      (i = (r = (t = e.publication) == null ? void 0 : t.isMuted) != null ? r : (n = e.participant.getTrackPublication(e.source)) == null ? void 0 : n.isMuted) != null ? i : !0
    )
  );
}
function rs(e) {
  return Ge(e, w.IsSpeakingChanged).pipe(
    P(([t]) => t)
  );
}
function is(e, t = {}) {
  var n;
  let r;
  const i = new k((c) => (r = c, () => a.unsubscribe())).pipe(R(Array.from(e.remoteParticipants.values()))), o = (n = t.additionalRoomEvents) != null ? n : Tn, s = Array.from(
    /* @__PURE__ */ new Set([
      g.ParticipantConnected,
      g.ParticipantDisconnected,
      g.ConnectionStateChanged,
      ...o
    ])
  ), a = ce(e, ...s).subscribe(
    (c) => r == null ? void 0 : r.next(Array.from(c.remoteParticipants.values()))
  );
  return e.remoteParticipants.size > 0 && (r == null || r.next(Array.from(e.remoteParticipants.values()))), i;
}
function os(e, t, n = {}) {
  var r;
  const i = (r = n.additionalEvents) != null ? r : En;
  return ce(
    e,
    g.ParticipantConnected,
    g.ParticipantDisconnected,
    g.ConnectionStateChanged
  ).pipe(
    xn((s) => {
      const a = s.getParticipantByIdentity(t);
      return a ? ye(a, ...i) : new k((c) => c.next(void 0));
    }),
    R(e.getParticipantByIdentity(t))
  );
}
function ss(e) {
  return Ge(
    e,
    w.ParticipantPermissionsChanged
  ).pipe(
    P(() => e.permissions),
    R(e.permissions)
  );
}
function as(e, { kind: t, identity: n }, r = {}) {
  var i;
  const o = (i = r.additionalEvents) != null ? i : En, s = (c) => {
    let u = !0;
    return t && (u = u && c.kind === t), n && (u = u && c.identity === n), u;
  };
  return ce(
    e,
    g.ParticipantConnected,
    g.ParticipantDisconnected,
    g.ConnectionStateChanged
  ).pipe(
    xn((c) => {
      const u = Array.from(c.remoteParticipants.values()).find(
        (l) => s(l)
      );
      return u ? ye(u, ...o) : new k((l) => l.next(void 0));
    }),
    R(Array.from(e.remoteParticipants.values()).find((c) => s(c)))
  );
}
function cs(e) {
  return typeof e > "u" ? new k() : Ge(e, w.AttributesChanged).pipe(
    P(([t]) => ({
      changed: t,
      attributes: e.attributes
    })),
    R({ changed: e.attributes, attributes: e.attributes })
  );
}
function us(e, t, n, r, i) {
  const { localParticipant: o } = t, s = (f, h) => {
    let d = !1;
    switch (f) {
      case D.Source.Camera:
        d = h.isCameraEnabled;
        break;
      case D.Source.Microphone:
        d = h.isMicrophoneEnabled;
        break;
      case D.Source.ScreenShare:
        d = h.isScreenShareEnabled;
        break;
    }
    return d;
  }, a = Ln(o).pipe(
    P((f) => s(e, f.participant)),
    R(s(e, o))
  ), c = new J(), u = (f, h) => V(this, null, function* () {
    try {
      switch (h ?? (h = n), c.next(!0), e) {
        case D.Source.Camera:
          return yield o.setCameraEnabled(
            f ?? !o.isCameraEnabled,
            h,
            r
          ), o.isCameraEnabled;
        case D.Source.Microphone:
          return yield o.setMicrophoneEnabled(
            f ?? !o.isMicrophoneEnabled,
            h,
            r
          ), o.isMicrophoneEnabled;
        case D.Source.ScreenShare:
          return yield o.setScreenShareEnabled(
            f ?? !o.isScreenShareEnabled,
            h,
            r
          ), o.isScreenShareEnabled;
        default:
          throw new TypeError("Tried to toggle unsupported source");
      }
    } catch (d) {
      if (i && d instanceof Error) {
        i == null || i(d);
        return;
      } else
        throw d;
    } finally {
      c.next(!1);
    }
  });
  return {
    className: U("button"),
    toggle: u,
    enabledObserver: a,
    pendingObserver: c.asObservable()
  };
}
function ls() {
  let e = !1;
  const t = new J(), n = new J(), r = (o) => V(this, null, function* () {
    n.next(!0), e = o ?? !e, t.next(e), n.next(!1);
  });
  return {
    className: U("button"),
    toggle: r,
    enabledObserver: t.asObservable(),
    pendingObserver: n.asObservable()
  };
}
function fs(e, t, n) {
  const r = new cn(void 0), i = so(t, e), o = (a, ...c) => V(this, [a, ...c], function* (u, l = {}) {
    var f, h, d;
    if (t) {
      _.debug(`Switching active device of kind "${e}" with id ${u}.`), yield t.switchActiveDevice(e, u, l.exact);
      const m = (f = t.getActiveDevice(e)) != null ? f : u;
      m !== u && u !== "default" && _.info(
        `We tried to select the device with id (${u}), but the browser decided to select the device with id (${m}) instead.`
      );
      let p;
      e === "audioinput" ? p = (h = t.localParticipant.getTrackPublication(D.Source.Microphone)) == null ? void 0 : h.track : e === "videoinput" && (p = (d = t.localParticipant.getTrackPublication(D.Source.Camera)) == null ? void 0 : d.track);
      const b = u === "default" && !p || u === "default" && (p == null ? void 0 : p.mediaStreamTrack.label.startsWith("Default"));
      r.next(b ? u : m);
    }
  });
  return {
    className: U("media-device-select"),
    activeDeviceObservable: i,
    setActiveMediaDevice: o
  };
}
function ds(e) {
  const t = (r) => {
    e.disconnect(r);
  };
  return { className: U("disconnect-button"), disconnect: t };
}
function ps(e) {
  const t = U("connection-quality"), n = co(e);
  return { className: t, connectionQualityObserver: n };
}
function hs(e) {
  let t = "track-muted-indicator-camera";
  switch (e.source) {
    case D.Source.Camera:
      t = "track-muted-indicator-camera";
      break;
    case D.Source.Microphone:
      t = "track-muted-indicator-microphone";
      break;
  }
  const n = U(t), r = uo(e);
  return { className: n, mediaMutedObserver: r };
}
function vs(e) {
  return { className: "lk-participant-name", infoObserver: ao(e) };
}
function ms() {
  return {
    className: U("participant-tile")
  };
}
var lo = {
  CHAT: "lk.chat"
}, fo = {
  CHAT: "lk-chat-topic"
};
function _n(e, t) {
  return V(this, arguments, function* (n, r, i = {}) {
    const { reliable: o, destinationIdentities: s, topic: a } = i;
    yield n.publishData(r, {
      destinationIdentities: s,
      topic: a,
      reliable: o
    });
  });
}
function po(e, t, n) {
  const r = Array.isArray(t) ? t : [t], i = ro(e).pipe(
    Ke(
      ([, , , c]) => t === void 0 || c !== void 0 && r.includes(c)
    ),
    P(([c, u, , l]) => {
      const f = {
        payload: c,
        topic: l,
        from: u
      };
      return n == null || n(f), f;
    })
  );
  let o;
  const s = new k((c) => {
    o = c;
  });
  return { messageObservable: i, isSendingObservable: s, send: (c, ...u) => V(this, [c, ...u], function* (l, f = {}) {
    o.next(!0);
    try {
      yield _n(e.localParticipant, l, X({ topic: r[0] }, f));
    } finally {
      o.next(!1);
    }
  }) };
}
var Le = /* @__PURE__ */ new WeakMap();
function ho(e) {
  return e.ignoreLegacy == !0;
}
var vo = (e) => JSON.parse(new TextDecoder().decode(e)), mo = (e) => new TextEncoder().encode(JSON.stringify(e));
function bs(e, t) {
  var n, r, i, o, s, a;
  const c = () => {
    var y, x, O;
    return ((y = e.serverInfo) == null ? void 0 : y.edition) === 1 || !!((x = e.serverInfo) != null && x.version) && Vn((O = e.serverInfo) == null ? void 0 : O.version, "1.8.2") > 0;
  }, u = new J(), l = (n = t == null ? void 0 : t.channelTopic) != null ? n : lo.CHAT, f = (r = t == null ? void 0 : t.channelTopic) != null ? r : fo.CHAT;
  let h = !1;
  Le.has(e) || (h = !0);
  const d = (i = Le.get(e)) != null ? i : /* @__PURE__ */ new Map(), m = (o = d.get(l)) != null ? o : new J();
  d.set(l, m), Le.set(e, d);
  const p = (s = t == null ? void 0 : t.messageDecoder) != null ? s : vo;
  if (h) {
    e.registerTextStreamHandler(l, (x, O) => V(this, null, function* () {
      const { id: $, timestamp: I } = x.info;
      Ye(x).pipe(
        st((T, L) => T + L),
        P((T) => ({
          id: $,
          timestamp: I,
          message: T,
          from: e.getParticipantByIdentity(O.identity)
          // editTimestamp: type === 'update' ? timestamp : undefined,
        }))
      ).subscribe({
        next: (T) => m.next(T)
      });
    }));
    const { messageObservable: y } = po(e, [f]);
    y.pipe(
      P((x) => {
        const O = p(x.payload);
        return ho(O) ? void 0 : Se(X({}, O), { from: x.from });
      }),
      Ke((x) => !!x),
      Dt(u)
    ).subscribe(m);
  }
  const b = m.pipe(
    st((y, x) => {
      if ("id" in x && y.find((O) => {
        var $, I;
        return (($ = O.from) == null ? void 0 : $.identity) === ((I = x.from) == null ? void 0 : I.identity) && O.id === x.id;
      })) {
        const O = y.findIndex(($) => $.id === x.id);
        if (O > -1) {
          const $ = y[O];
          y[O] = Se(X({}, x), {
            timestamp: $.timestamp,
            editTimestamp: x.timestamp
          });
        }
        return [...y];
      }
      return [...y, x];
    }, []),
    Dt(u)
  ), v = new cn(!1), S = (a = t == null ? void 0 : t.messageEncoder) != null ? a : mo, E = (y, x) => V(this, null, function* () {
    var O;
    x || (x = {}), (O = x.topic) != null || (x.topic = l), v.next(!0);
    try {
      const I = {
        id: (yield e.localParticipant.sendText(y, x)).id,
        timestamp: Date.now(),
        message: y,
        from: e.localParticipant,
        attachedFiles: x.attachments
      };
      m.next(I);
      const z = S(Se(X({}, I), {
        ignoreLegacy: c()
      }));
      return yield _n(e.localParticipant, z, {
        reliable: !0,
        topic: f
      }), I;
    } finally {
      v.next(!1);
    }
  });
  function A() {
    u.next(), u.complete(), m.complete(), Le.delete(e), e.unregisterTextStreamHandler(l);
  }
  return e.once(g.Disconnected, A), {
    messageObservable: b,
    isSendingObservable: v,
    send: E
  };
}
function gs() {
  const e = (n) => V(this, null, function* () {
    _.info("Start Audio for room: ", n), yield n.startAudio();
  });
  return { className: U("start-audio-button"), roomAudioPlaybackAllowedObservable: io, handleStartAudioPlayback: e };
}
function ys() {
  const e = (n) => V(this, null, function* () {
    _.info("Start Video for room: ", n), yield n.startVideo();
  });
  return { className: U("start-audio-button"), roomVideoPlaybackAllowedObservable: oo, handleStartVideoPlayback: e };
}
function ws() {
  return { className: [U("button"), U("chat-toggle")].join(" ") };
}
function xs() {
  return { className: [U("button"), U("focus-toggle-button")].join(" ") };
}
function Ss() {
  return { className: "lk-clear-pin-button lk-button" };
}
function Ts() {
  return { className: "lk-room-container" };
}
function jt(e, t, n = !0) {
  const i = [e.localParticipant, ...Array.from(e.remoteParticipants.values())], o = [];
  return i.forEach((s) => {
    t.forEach((a) => {
      const c = Array.from(
        s.trackPublications.values()
      ).filter(
        (u) => u.source === a && // either return all or only the ones that are subscribed
        (!n || u.track)
      ).map((u) => ({
        participant: s,
        publication: u,
        source: u.source
      }));
      o.push(...c);
    });
  }), { trackReferences: o, participants: i };
}
function Wt(e, t, n = !1) {
  const { sources: r, kind: i, name: o } = t;
  return Array.from(e.trackPublications.values()).filter(
    (a) => (!r || r.includes(a.source)) && (!i || a.kind === i) && (!o || a.trackName === o) && // either return all or only the ones that are subscribed
    (!n || a.track)
  ).map((a) => ({
    participant: e,
    publication: a,
    source: a.source
  }));
}
function Es(e, t, n) {
  var r, i;
  const o = (r = n.additionalRoomEvents) != null ? r : Tn, s = (i = n.onlySubscribed) != null ? i : !0, a = Array.from(
    (/* @__PURE__ */ new Set([
      g.ParticipantConnected,
      g.ParticipantDisconnected,
      g.ConnectionStateChanged,
      g.LocalTrackPublished,
      g.LocalTrackUnpublished,
      g.TrackPublished,
      g.TrackUnpublished,
      g.TrackSubscriptionStatusChanged,
      ...o
    ])).values()
  );
  return ce(e, ...a).pipe(
    P((u) => {
      const l = jt(u, t, s);
      return _.debug(`TrackReference[] was updated. (length ${l.trackReferences.length})`, l), l;
    }),
    R(jt(e, t, s))
  );
}
function Cs(e, t) {
  return ye(e, ...Wi).pipe(
    P((r) => {
      const i = Wt(r, t);
      return _.debug(`TrackReference[] was updated. (length ${i.length})`, i), i;
    }),
    R(Wt(e, t))
  );
}
function In(e, t) {
  return new k((r) => {
    const i = (...s) => {
      r.next(s);
    };
    return e.on(t, i), () => {
      e.off(t, i);
    };
  });
}
function Ps(e) {
  return In(e, Vt.TranscriptionReceived);
}
function As(e) {
  return In(e, Vt.TimeSyncUpdate).pipe(
    P(([t]) => t)
  );
}
function Os(e, t = 1e3) {
  if (e === null) return It(!1);
  const n = ot(e, "mousemove", { passive: !0 }).pipe(P(() => !0)), r = n.pipe(
    li({
      each: t,
      with: () => Ue(It(!1), r.pipe(ki(n)))
    }),
    Ci()
  );
  return r;
}
function bo(e, t) {
  if (typeof localStorage > "u") {
    _.error("Local storage is not available.");
    return;
  }
  try {
    if (t) {
      const n = Object.fromEntries(
        Object.entries(t).filter(([, r]) => r !== "")
      );
      localStorage.setItem(e, JSON.stringify(n));
    }
  } catch (n) {
    _.error(`Error setting item to local storage: ${n}`);
  }
}
function go(e) {
  if (typeof localStorage > "u") {
    _.error("Local storage is not available.");
    return;
  }
  try {
    const t = localStorage.getItem(e);
    if (!t) {
      _.warn(`Item with key ${e} does not exist in local storage.`);
      return;
    }
    return JSON.parse(t);
  } catch (t) {
    _.error(`Error getting item from local storage: ${t}`);
    return;
  }
}
function yo(e) {
  return {
    load: () => go(e),
    save: (t) => bo(e, t)
  };
}
var wo = `${Sn}-user-choices`, xe = {
  videoEnabled: !0,
  audioEnabled: !0,
  videoDeviceId: "default",
  audioDeviceId: "default",
  username: ""
}, { load: xo, save: So } = yo(wo);
function ks(e, t = !1) {
  t !== !0 && So(e);
}
function Ls(e, t = !1) {
  var n, r, i, o, s;
  const a = {
    videoEnabled: (n = e == null ? void 0 : e.videoEnabled) != null ? n : xe.videoEnabled,
    audioEnabled: (r = e == null ? void 0 : e.audioEnabled) != null ? r : xe.audioEnabled,
    videoDeviceId: (i = e == null ? void 0 : e.videoDeviceId) != null ? i : xe.videoDeviceId,
    audioDeviceId: (o = e == null ? void 0 : e.audioDeviceId) != null ? o : xe.audioDeviceId,
    username: (s = e == null ? void 0 : e.username) != null ? s : xe.username
  };
  if (t)
    return a;
  {
    const c = xo();
    return X(X({}, a), c ?? {});
  }
}
var Ze = null, et = null, To = 0;
function Bt() {
  return Ze || (Ze = /* @__PURE__ */ new Map()), Ze;
}
function Eo() {
  return et || (et = /* @__PURE__ */ new WeakMap()), et;
}
function Co(e, t) {
  const n = Eo();
  let r = n.get(e);
  return r || (r = `room_${To++}`, n.set(e, r)), `${r}:${t}`;
}
function _s(e, t) {
  const n = Co(e, t), r = Bt(), i = r.get(n);
  if (i)
    return i;
  const o = new J(), s = [];
  e.registerTextStreamHandler(t, (c, u) => V(this, null, function* () {
    Ye(c).pipe(
      st((f, h) => f + h, "")
    ).subscribe((f) => {
      const h = s.findIndex((d) => d.streamInfo.id === c.info.id);
      h !== -1 ? (s[h] = Se(X({}, s[h]), {
        text: f
      }), o.next([...s])) : (s.push({
        text: f,
        participantInfo: u,
        streamInfo: c.info
      }), o.next([...s]));
    });
  }));
  const a = o.asObservable().pipe(Oi());
  return r.set(n, a), e.once(g.Disconnected, () => {
    o.complete(), Bt().delete(n);
  }), a;
}
function Mn(e, t) {
  if (t.msg === "show_chat")
    return { ...e, showChat: !0, unreadMessages: 0 };
  if (t.msg === "hide_chat")
    return { ...e, showChat: !1 };
  if (t.msg === "toggle_chat") {
    const n = { ...e, showChat: !e.showChat };
    return n.showChat === !0 && (n.unreadMessages = 0), n;
  } else return t.msg === "unread_msg" ? { ...e, unreadMessages: t.count } : t.msg === "toggle_settings" ? { ...e, showSettings: !e.showSettings } : { ...e };
}
function Dn(e, t) {
  return t.msg === "set_pin" ? [t.trackReference] : t.msg === "clear_pin" ? [] : { ...e };
}
const Rn = M.createContext(void 0);
function Is() {
  const e = M.useContext(Rn);
  if (!e)
    throw Error("Tried to access LayoutContext context outside a LayoutContextProvider provider.");
  return e;
}
function Ms(e) {
  const t = Po();
  if (e ?? (e = t), !e)
    throw Error("Tried to access LayoutContext context outside a LayoutContextProvider provider.");
  return e;
}
function Ds() {
  const [e, t] = M.useReducer(Dn, Cn), [n, r] = M.useReducer(Mn, Pn);
  return {
    pin: { dispatch: t, state: e },
    widget: { dispatch: r, state: n }
  };
}
function Rs(e) {
  const [t, n] = M.useReducer(Dn, Cn), [r, i] = M.useReducer(Mn, Pn);
  return e ?? {
    pin: { dispatch: n, state: t },
    widget: { dispatch: i, state: r }
  };
}
function Po() {
  return M.useContext(Rn);
}
const $n = M.createContext(
  void 0
);
function $s() {
  const e = M.useContext($n);
  if (!e)
    throw Error("tried to access track context outside of track context provider");
  return e;
}
function Nn() {
  return M.useContext($n);
}
function Ns(e) {
  const t = Nn(), n = e ?? t;
  if (!n)
    throw new Error(
      "No TrackRef, make sure you are inside a TrackRefContext or pass the TrackRef explicitly"
    );
  return n;
}
const Un = M.createContext(void 0);
function Us() {
  const e = M.useContext(Un);
  if (!e)
    throw Error("tried to access participant context outside of participant context provider");
  return e;
}
function Ao() {
  return M.useContext(Un);
}
function Fs(e) {
  const t = Ao(), n = Nn(), r = e ?? t ?? (n == null ? void 0 : n.participant);
  if (!r)
    throw new Error(
      "No participant provided, make sure you are inside a participant context or pass the participant explicitly"
    );
  return r;
}
const Fn = M.createContext(void 0);
function js() {
  const e = M.useContext(Fn);
  if (!e)
    throw Error("tried to access room context outside of livekit room component");
  return e;
}
function Oo() {
  return M.useContext(Fn);
}
function Ws(e) {
  const t = Oo(), n = e ?? t;
  if (!n)
    throw new Error(
      "No room provided, make sure you are inside a Room context or pass the room explicitly"
    );
  return n;
}
const ko = M.createContext(void 0);
function Bs(e) {
  const t = M.useContext(ko);
  if (e === !0) {
    if (t)
      return t;
    throw Error("tried to access feature context, but none is present");
  }
  return t;
}
export {
  ks as $,
  qo as A,
  Ao as B,
  ao as C,
  ms as D,
  is as E,
  Ms as F,
  Bo as G,
  os as H,
  as as I,
  Xo as J,
  Zo as K,
  ko as L,
  Go as M,
  gs as N,
  ys as O,
  ws as P,
  hs as Q,
  Fn as R,
  us as S,
  ls as T,
  zi as U,
  Es as V,
  Yo as W,
  Ft as X,
  Qo as Y,
  bs as Z,
  Ls as _,
  Is as a,
  ts as a0,
  W as a1,
  Cs as a2,
  As as a3,
  Ps as a4,
  zo as a5,
  Ho as a6,
  cs as a7,
  ns as a8,
  _s as a9,
  Nn as aa,
  Rn as ab,
  vs as ac,
  _o as ad,
  Tr as ae,
  Un as af,
  $n as ag,
  Bs as ah,
  Os as ai,
  Do as aj,
  Rs as ak,
  Fo as al,
  Uo as am,
  $o as an,
  No as ao,
  Vo as ap,
  Ds as aq,
  Io as ar,
  Ro as as,
  jo as at,
  Wo as au,
  Us as av,
  $s as aw,
  Ss as b,
  Fs as c,
  ps as d,
  Jo as e,
  js as f,
  po as g,
  ds as h,
  Ns as i,
  Po as j,
  xs as k,
  _ as l,
  Mo as m,
  Vi as n,
  uo as o,
  N as p,
  rs as q,
  io as r,
  Ts as s,
  Ln as t,
  Ws as u,
  ss as v,
  Oo as w,
  es as x,
  fs as y,
  Ko as z
};
